---
title: "Differential gene expression analysis between sample `r params$sample1` vs. `r params$sample2`"
author:
  - "trnguyen@ukaachen.de"
date: "Last update on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    keep_md: no
    df_print: paged
    toc: true
    toc_float:
      toc_collapsed: false
    toc_depth: 3
    number_sections: true
    theme: lumen
params:
  sample1: NA
  sample2: NA
  path.to.s.obj: NA
  path.to.save.output: NA
  cluster.name: NA
  condition.name: NA
  reduction.name: NA
---


```{css zoom-lib-src, echo = FALSE}
script src = "https://ajax.googleapis.com/ajax/libs/jquery/3.4.1/jquery.min.js"
```

```{js zoom-jquery, echo = FALSE}
 $(document).ready(function() {
    $('body').prepend('<div class=\"zoomDiv\"><img src=\"\" class=\"zoomImg\"></div>');
    // onClick function for all plots (img's)
    $('img:not(.zoomImg)').click(function() {
      $('.zoomImg').attr('src', $(this).attr('src')).css({width: '100%'});
      $('.zoomDiv').css({opacity: '1', width: 'auto', border: '1px solid white', borderRadius: '5px', position: 'fixed', top: '50%', left: '50%', marginRight: '-50%', transform: 'translate(-50%, -50%)', boxShadow: '0px 0px 50px #888888', zIndex: '50', overflow: 'auto', maxHeight: '100%'});
    });
    // onClick function for zoomImg
    $('img.zoomImg').click(function() {
      $('.zoomDiv').css({opacity: '0', width: '0%'}); 
    });
  });
```

<style type="text/css">
    div.datatables { height: auto !important;}
</style>

# Differential gene expression for each cluster of cells

- **positive** `logFC` indicates that the gene is highly expressed in the sample 1, `r params$sample1`.

- **negative** `logFC` indicates that the gene is highly expressed in the sample 2, `r params$sample2`.


```{r echo=FALSE, warning=FALSE, message=FALSE, results='hide', fig.width=20, fig.height=12}
#####----------------------------------------------------------------------#####
#
# trnguyen@ukaachen.de
#
#####----------------------------------------------------------------------#####

##### clean up #####
# gc()
# rm(list = ls())
if ("org.Mm.eg.db" %in% installed.packages() == FALSE){
  BiocManager::install("org.Mm.eg.db", update = FALSE)
}

if ("heatmaply" %in% installed.packages() == FALSE){
  install.packages("heatmaply")
}

library(org.Mm.eg.db)

#####----------------------------------------------------------------------#####
scrna_pipeline_src <- "/home/hieunguyen/CRC1382/src_2023/src_pipeline/scRNA_GEX_pipeline_SeuratV5/processes_src"
source(file.path(scrna_pipeline_src, "import_libraries.R"))
source(file.path(scrna_pipeline_src, "helper_functions.R"))
# 
sample1 <- params$sample1
sample2 <- params$sample2
path.to.s.obj <- params$path.to.s.obj
path.to.save.output <- params$path.to.save.output
cluster.name <- params$cluster.name
condition.name <- params$condition.name
reduction.name <- params$reduction.name

# sample1 <- "treated_gut_CD45"
# sample2 <- "ctrl_gut_CD45"
# path.to.s.obj <- "/media/hieunguyen/HD01/outdir/CRC1382/SBharadwaj_20250102/SBharadwaj_20240318_Sample_1_4_7_8_2_5/data_analysis/07_output/preprocees_sub_clusters/v0.1/DC/noReIntegration/Project_SBharadwaj_20240318_Sample_1_4_7_8_2_5_v0.1_DC.addedInfo.rds"
# path.to.save.output <- "/home/hieunguyen/CRC1382/outdir/tmp_output"
# cluster.name <- "seurat_clusters"
# condition.name <- "condition"
# reduction.name <- "SCT_UMAP"


dir.create(path.to.save.output, showWarnings = FALSE, recursive = TRUE)
library(clusterProfiler)
if (packageVersion("org.Mm.eg.db") != "3.19.1"){
  install.packages("https://bioconductor.org/packages/release/bioc/src/contrib/AnnotationDbi_1.66.0.tar.gz", 
                   type = "source", repos = NULL)
  install.packages("https://bioconductor.org/packages/release/data/annotation/src/contrib/org.Mm.eg.db_3.19.1.tar.gz", 
                   type = "source", repos = NULL)
}
library(org.Mm.eg.db)
```

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
print(sprintf("Path to the input Seurat object: %s", path.to.s.obj))
print(sprintf("Sample 1: %s", sample1))
print(sprintf("Sample 2: %s", sample2))
```

```{r echo=FALSE, warning=FALSE, message=FALSE, results='hide', fig.width=20, fig.height=12}
s.obj <- readRDS(path.to.s.obj)
s.obj.original <- s.obj
```

# Dataset overview

## All samples

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=14, fig.height=10}
DimPlot(object = s.obj, reduction = reduction.name, label = TRUE, label.box = TRUE, repel = TRUE)
```

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=14, fig.height=10}
DimPlot(object = s.obj, reduction = reduction.name, label = TRUE, label.box = TRUE, repel = TRUE, split.by = condition.name)
```

```{r echo=FALSE, warning=FALSE, message=FALSE, results='hide', fig.width=20, fig.height=12}
all.genes.in.Mm.db <- AnnotationDbi::keys(org.Mm.eg.db, keytype = "SYMBOL")
```

**In this analysis, we conduct the Differential gene expression analysis between sample** `r params$sample1` vs. `r params$sample2`  **in the dataset ** `r basename(params$path.to.s.obj)`**. We also removed TCR genes.**

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=14, fig.height=10}
TR_genes_patterns <- c("Trav", "Traj", "Trac", "Trbv", "Trbd", "Trbj", "Trbc",
                             "Trgv", "Trgj", "Trgc", "Trdv", "Trdc", "Trdj") 
DefaultAssay(s.obj) <- "SCT"

all_genes <- row.names(s.obj)
      
genes.to.exclude <- unlist(lapply(all_genes, function(x){
        if (substr(x, 1, 4) %in% TR_genes_patterns){
          return(x)
        } else {
          return(NA)
        }
}))
genes.to.exclude <- subset(genes.to.exclude, is.na(genes.to.exclude) == FALSE)
genes.to.keep <- setdiff(row.names(s.obj), genes.to.exclude)
```

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=14, fig.height=10}
if (condition.name == "condition"){
  s.obj <- subset(s.obj, condition %in% c(sample1, sample2) )  
} else {
  s.obj <- subset(s.obj, global.condition %in% c(sample1, sample2) )
}
```

## Re-plot the UMAP

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=14, fig.height=10}
DimPlot(object = s.obj, reduction = reduction.name, label = TRUE, label.box = TRUE, repel = TRUE)
```

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=14, fig.height=10}
DimPlot(object = s.obj, reduction = reduction.name, label = TRUE, label.box = TRUE, repel = TRUE, split.by = condition.name)
```

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=14, fig.height=10}
savename <- sprintf("sample_%s_vs_%s.single_cell.rds", sample1, sample2)
savename.raw <- sprintf("sample_%s_vs_%s.raw.single_cell.rds", sample1, sample2)

if (cluster.name == "cca.cluster.0.5"){
  cell.metadata <- s.obj@meta.data %>% rownames_to_column("barcode") %>% subset(select = c(barcode, cca.cluster.0.5)) 
} else if (cluster.name == "seurat_clusters"){
  cell.metadata <- s.obj@meta.data %>% rownames_to_column("barcode") %>% subset(select = c(barcode, seurat_clusters)) 
} else {
  cell.metadata <- s.obj@meta.data %>% rownames_to_column("barcode") %>% subset(select = c(barcode, cell.annotation))
}
```



```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=14, fig.height=10}
if (file.exists(file.path(path.to.save.output, savename)) == FALSE){
  de.markers <- hash()
  de.markers.raw <- hash()
  num.clusters <- length(unique(s.obj@meta.data[[cluster.name]]))
  
  for (cluster.id in unique(s.obj@meta.data[[cluster.name]])){
    selected.cells <- subset(cell.metadata, cell.metadata[[cluster.name]] == cluster.id)$barcode
    tmp.s.obj <- subset(s.obj, cells = selected.cells)
    if (condition.name == "condition"){
      count.cells <- tmp.s.obj@meta.data %>% subset(select = c(condition)) %>% table %>% data.frame()      
    } else {
      count.cells <- tmp.s.obj@meta.data %>% subset(select = c(global.condition)) %>% table %>% data.frame()
    }

    if (sample1 %in% count.cells[[condition.name]] == FALSE | sample2 %in% count.cells[[condition.name]] == FALSE){
      if (sample1 %in% count.cells[[condition.name]] == FALSE){
        count.sample1 <- 0
        count.sample2 <- subset(count.cells, count.cells[[condition.name]] == sample2)$Freq
      }
      if (sample2 %in% count.cells[[condition.name]] == FALSE){
        count.sample1 <- subset(count.cells, count.cells[[condition.name]] == sample1)$Freq
        count.sample2 <- 0
      }
    } else {
      count.sample1 <- subset(count.cells, count.cells[[condition.name]] == sample1)$Freq
      count.sample2 <- subset(count.cells, count.cells[[condition.name]] == sample2)$Freq
    }
    
    if (count.sample1 >= 10 & count.sample2 >= 10){
      tmp.s.obj <- PrepSCTFindMarkers(tmp.s.obj)
      tmp.de.markers <- FindMarkers(object = tmp.s.obj, 
                                    ident.1 = sample1, 
                                    ident.2 = sample2,
                                    group.by = condition.name, 
                                    test.use = "wilcox", 
                                    assay = "SCT",
                                    features = intersect(row.names(s.obj), genes.to.keep), 
                                    recorrect_umi = FALSE)
      
      tmp.de.markers.raw <-  tmp.de.markers %>% 
          rownames_to_column("Gene") %>%
          rowwise() %>%
          mutate(abs_avg_log2FC = abs(avg_log2FC)) %>%
          arrange(desc(abs_avg_log2FC))
      
      tmp.de.markers <- subset(tmp.de.markers, tmp.de.markers$p_val_adj <= 0.05) %>% 
          rownames_to_column("Gene") %>%
          rowwise() %>%
          mutate(abs_avg_log2FC = abs(avg_log2FC)) %>%
          arrange(desc(abs_avg_log2FC))
    } else {
      tmp.de.markers <- data.frame(data = c("Not enough cells to conduct DE test, either sample1 <= 10 cells or sample2 <= 10 cells"))
      tmp.de.markers.raw <- data.frame(data = c("Not enough cells to conduct DE test, either sample1 <= 10 cells or sample2 <= 10 cells"))
    }
    if (cluster.name %in% c("cca.cluster.0.5", "seurat_clusters")){
      de.markers[[sprintf("cluster_%s", cluster.id)]] <- tmp.de.markers
      de.markers.raw[[sprintf("cluster_%s", cluster.id)]] <- tmp.de.markers.raw      
    } else {
      de.markers[[cluster.id]] <- tmp.de.markers
      de.markers.raw[[cluster.id]] <- tmp.de.markers.raw
    }
  }
  saveRDS(object = de.markers, file.path(path.to.save.output, savename))
  saveRDS(object = de.markers.raw, file.path(path.to.save.output, savename.raw))
} else {
  de.markers <- readRDS(file.path(path.to.save.output, savename))
  de.markers.raw <- readRDS(file.path(path.to.save.output, savename.raw))
}
```

## Full tables of all D.E genes between sample `r params$sample1` vs `r params$sample2`` {.tabset}
```{r echo=FALSE, include=FALSE}
de.markers[[sprintf("cluster_%s", 0)]] %>% create_dt()
```

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
if (cluster.name %in%  c("cca.cluster.0.5", "seurat_clusters")){
  all.clusters.in.de.markers <- sort(to_vec( for (item in names(de.markers)) as.numeric(str_split(item, "_")[[1]][[2]]) ))
  all.clusters.in.de.markers <- to_vec( for(item in all.clusters.in.de.markers) sprintf("cluster_%s", item))  
} else {
  all.clusters.in.de.markers <- names(de.markers)
}

for (cluster.id in all.clusters.in.de.markers){
  tmp <- de.markers[[cluster.id]] %>% mutate_if(is.numeric, round, 5) 
  cat(paste("\n\n###", cluster.id, "##\n"))
  print( htmltools::tagList(datatable(tmp, extensions = 'Buttons',
                filter = "top",
                options = list(dom = 'Blfrtip',
                               buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                               lengthMenu = list(c(10,25,50,-1),
                                                 c(10,25,50,"All")),
                               columnDefs = list(list(
                                 targets = "_all",
                                 render = JS(
                                   "function(data, type, row, meta) {",
                                   "return type === 'display' && data != null && data.length > 100 ?",
                                   "'<span title=\"' + data + '\">' + data.substr(0, 100) + '...</span>' : data;",
                                   "}")
                               ))
                ))))
  cat("\n \n")
}
```

## Volcano plot {.tabset}

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=14, fig.height=10}
for (cluster.id in all.clusters.in.de.markers){
  cat(sprintf("### Cluster %s \n", cluster.id))
  input.df <- de.markers.raw[[cluster.id]] 
  if ("data" %in% colnames(input.df) == FALSE & nrow(input.df) != 0){
    top10up.genes <- input.df %>% subset(p_val_adj <= 0.05) %>% arrange(desc(avg_log2FC)) %>% head(10)
    top10down.genes <- input.df %>% subset(p_val_adj <= 0.05) %>% arrange(desc(avg_log2FC)) %>% tail(10)
    input.df <- input.df %>%
    rowwise() %>%
    mutate(sig = ifelse(p_val_adj <= 0.05, "Sig.", "not Sig.")) %>%
      mutate(show.gene.name = ifelse(Gene %in% c(top10up.genes$Gene, top10down.genes$Gene), Gene, NA))
      
    
    cutoff.adjp <- 0.05
    volcano.plot <- ggplot(data=input.df, 
                           aes(x=avg_log2FC, y=-log10(p_val_adj), col = sig, label=Gene)) + 
      geom_point(size = 3) + geom_label_repel(label = input.df$show.gene.name, size = 8) + 
      scale_color_manual(values=c("#c0d2f0", "#f28095")) +
      geom_vline(xintercept=c(-1, 1), col="#9a9fa6") +
      geom_hline(yintercept=-log10(cutoff.adjp), col="#9a9fa6") +
      theme(plot.title = element_text(hjust=0.5, face="bold", size = 12), axis.text=element_text(size=12)) + 
      theme_bw()
    
  } else {
    volcano.plot <- ggplot() + ggtitle("Not enough cells to conduct DE test")
  }
  print(volcano.plot)
  cat("\n \n")
}

```

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
library("clusterProfiler")
```

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
diff.markers.df <- data.frame()
for (cluster.id in names(de.markers.raw)){
  tmp <- de.markers.raw[[cluster.id]] %>% mutate(cluster = cluster.id)
  if ("data" %in% colnames(tmp) == FALSE){
    diff.markers.df <- rbind(diff.markers.df, tmp)    
  }
}
if ("org.Mm.eg.db" %in% installed.packages() == FALSE){
  BiocManager::install("org.Mm.eg.db")
}
library("org.Mm.eg.db")
```

## Heatmaps {.tabset}
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=15}
mapal <- colorRampPalette(RColorBrewer::brewer.pal(11,"RdBu"))(256)
for (cluster.id in all.clusters.in.de.markers){
  cat(sprintf("### Cluster %s \n", cluster.id))
  inputdf <- de.markers[[cluster.id]] 
  if ("data" %in% colnames(inputdf) | nrow(inputdf) == 0){
    p <- ggplot() + ggtitle("No significantly DE genes found or not enough cells to conduct DE test!")
  } else {
    top10up.genes <- inputdf %>% subset(p_val_adj <= 0.05 & avg_log2FC > 0) %>% arrange(desc(avg_log2FC)) %>% head(20)
    top10down.genes <- inputdf %>% subset(p_val_adj <= 0.05 & avg_log2FC < 0) %>% arrange(desc(avg_log2FC)) %>% tail(20)
    
    if (length(c(top10up.genes$Gene, top10down.genes$Gene)) > 1){
      if (cluster.name == "cca.cluster.0.5"){
          p <- DoHeatmap(object = subset(s.obj, cca.cluster.0.5 == as.numeric(str_replace(cluster.id, "cluster_", ""))), 
                     features = c(top10up.genes$Gene, top10down.genes$Gene), group.by = condition.name, assay = "SCT") + 
        scale_fill_gradientn(colours = rev(mapal))
      } else if (cluster.name == "seurat_clusters"){
        p <- DoHeatmap(object = subset(s.obj, seurat_clusters == as.numeric(str_replace(cluster.id, "cluster_", ""))), 
                     features = c(top10up.genes$Gene, top10down.genes$Gene), group.by = condition.name, assay = "SCT") + 
        scale_fill_gradientn(colours = rev(mapal))
      }else if (cluster.name == "cell.annotation") {
          p <- DoHeatmap(object = subset(s.obj, cell.annotation == cluster.id ), 
                     features = c(top10up.genes$Gene, top10down.genes$Gene), group.by = condition.name, assay = "SCT") + 
        scale_fill_gradientn(colours = rev(mapal))
      }

    } else {
      p <- ggplot() + ggtitle("No significantly DE genes found or not enough cells to conduct DE test!")
    }
  }
  print(p)
  cat("\n \n")
}
```

# Pathway analysis for the cluster specific DE test (single cell data)
## Full list of genes ranked by `avg_log2FC` {.tabset}

```{r echo=FALSE, warning=FALSE, results='asis', include=FALSE}
diff.markers.df <- data.frame()
for (cluster.id in names(de.markers.raw)){
  tmp <- de.markers.raw[[cluster.id]] %>% mutate(cluster = cluster.id)
  if ("data" %in% colnames(tmp) == FALSE){
    diff.markers.df <- rbind(diff.markers.df, tmp)    
  }
}
```

```{r echo=FALSE, warning=FALSE, results='asis', include=FALSE}
diff.markers.df %>% create_dt()
```

```{r echo=FALSE, warning=FALSE, results='asis'}
cat("List of significantly differentially expressed genes between 2 samples \n")

all.cluster.ids <- unique(diff.markers.df$cluster)

all.cluster.ids.idx <- sort(to_vec(for (item in all.cluster.ids)
  as.numeric(str_replace(item, "cluster_", ""))))
all.cluster.ids <- to_vec(for (item in all.cluster.ids.idx)
  sprintf("cluster_%s", item))

for(cluster.id in all.cluster.ids){
  df.cluster.id <- subset(diff.markers.df, diff.markers.df$cluster == cluster.id) %>% mutate_if(is.numeric, round, 6)
  cat(paste("\n\n### Cluster: ", cluster.id, "##\n"))
  print( htmltools::tagList(datatable(df.cluster.id, extensions = 'Buttons',
                filter = "top",
                options = list(dom = 'Blfrtip',
                               buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                               lengthMenu = list(c(10,25,50,-1),
                                                 c(10,25,50,"All")),
                               columnDefs = list(list(
                                 targets = "_all",
                                 render = JS(
                                   "function(data, type, row, meta) {",
                                   "return type === 'display' && data != null && data.length > 100 ?",
                                   "'<span title=\"' + data + '\">' + data.substr(0, 100) + '...</span>' : data;",
                                   "}")
                               ))
                ))))
  cat("\n \n")
}

```

## Genes of interest (DE genes only) {.tabset } 
```{r echo=FALSE, warning=FALSE, results='asis', include=FALSE}
diff.markers.df %>% create_dt()
```

```{r echo=FALSE, warning=FALSE, results='asis'}

for(cluster.id in all.cluster.ids){
  df.cluster.id <- subset(diff.markers.df, diff.markers.df$cluster == cluster.id & diff.markers.df$p_val_adj <= 0.05) %>% mutate_if(is.numeric, round, 6)
  cat(paste("\n\n### Cluster: ", cluster.id, "##\n"))
    print( htmltools::tagList(datatable(df.cluster.id, extensions = 'Buttons',
                filter = "top",
                options = list(dom = 'Blfrtip',
                               buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                               lengthMenu = list(c(10,25,50,-1),
                                                 c(10,25,50,"All")),
                               columnDefs = list(list(
                                 targets = "_all",
                                 render = JS(
                                   "function(data, type, row, meta) {",
                                   "return type === 'display' && data != null && data.length > 100 ?",
                                   "'<span title=\"' + data + '\">' + data.substr(0, 100) + '...</span>' : data;",
                                   "}")
                               ))
                ))))
  cat("\n \n")
}
```

## Over-representation analyiss (ORA)

### GO-ORA {.tabset}
```{r echo=FALSE, warning=FALSE, results='asis'}
if (file.exists(file.path(path.to.save.output, "all_ORA_GO_objects.single_cell.rds")) == FALSE){
  go.ora <- hash()
  all.ORA.GO.objects <- hash()
  for (cluster.id in all.cluster.ids){
    input.gene.list <- de.markers[[cluster.id]]$Gene
    if (length(input.gene.list) != 0){
      ora.GO <- enrichGO(gene = input.gene.list,
                        OrgDb = org.Mm.eg.db,
                        ont = "ALL",
                        pvalueCutoff = 0.05,
                        qvalueCutoff = 0.05,
                        readable = TRUE,
                        keyType = "SYMBOL",
                        pAdjustMethod = "BH")
      
      ora.GOdf <- as.data.frame(ora.GO)
      if (nrow(ora.GOdf) == 0){
        go.ora[[cluster.id]] <- data.frame(data = c("No DE genes to conduct ORA  or no pathway found"))
      } else {
        ora.GOdf <- ora.GOdf[order(ora.GOdf$p.adjust, decreasing = FALSE),]  %>% mutate_if(is.numeric, round, 6) 
        all.ORA.GO.objects[[cluster.id]] <- ora.GO
        go.ora[[cluster.id]] <- ora.GOdf
      }
    } else {
      go.ora[[cluster.id]] <- data.frame(data = c("No DE genes to conduct ORA"))
    }
  }
  saveRDS(all.ORA.GO.objects, file.path(path.to.save.output, "all_ORA_GO_objects.single_cell.rds"))
  saveRDS(go.ora, file.path(path.to.save.output, "all_ORA_GO_tables.single_cell.rds"))
} else {
  all.ORA.GO.objects <- readRDS(file.path(path.to.save.output, "all_ORA_GO_objects.single_cell.rds"))
  go.ora <- readRDS(file.path(path.to.save.output, "all_ORA_GO_tables.single_cell.rds"))
}
```

```{r echo=FALSE, warning=FALSE, results='asis', include=FALSE}
go.ora[["cluster_0"]] %>% create_dt()
```

```{r echo=FALSE, warning=FALSE, results='asis'}
for (cluster.id in names(go.ora)){
  cat(sprintf("#### %s \n", cluster.id))
  tmp <- go.ora[[cluster.id]]
  print( htmltools::tagList(datatable(tmp, extensions = 'Buttons',
                filter = "top",
                options = list(dom = 'Blfrtip',
                               buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                               lengthMenu = list(c(10,25,50,-1),
                                                 c(10,25,50,"All")),
                               columnDefs = list(list(
                                 targets = "_all",
                                 render = JS(
                                   "function(data, type, row, meta) {",
                                   "return type === 'display' && data != null && data.length > 100 ?",
                                   "'<span title=\"' + data + '\">' + data.substr(0, 100) + '...</span>' : data;",
                                   "}")
                               ))
                ))))
  cat("\n \n")
}
```

### KEGG-ORA {.tabset}
```{r echo=FALSE, warning=FALSE, results='asis', message=FALSE}
if (file.exists(file.path(path.to.save.output, "all_ORA_KEGG_objects.single_cell.rds")) == FALSE){
  kegg.ora <- hash()
  all.ORA.KEGG.objects <- hash()
  for (cluster.id in all.cluster.ids){
    input.gene.list <- de.markers[[cluster.id]]$Gene 
    if (length(input.gene.list) != 0) {
      if (length(intersect(input.gene.list, all.genes.in.Mm.db)) != 0){
        input.gene.list <- bitr(input.gene.list, fromType="SYMBOL", toType="ENTREZID", OrgDb="org.Mm.eg.db")
        ora.KEGG <-  enrichKEGG(gene = input.gene.list$ENTREZID,
                         organism     = 'mmu',
                         pvalueCutoff = 0.05, 
                         qvalueCutoff = 0.05)
        ora.KEGGdf <- as.data.frame(ora.KEGG)
        if (nrow(ora.KEGGdf) == 0){
          kegg.ora[[cluster.id]] <- data.frame(data = c("No DE genes to conduct ORA  or no pathway found"))
        } else {
                ora.KEGGdf <- ora.KEGGdf[order(ora.KEGGdf$p.adjust), ] %>% mutate_if(is.numeric, round, 6) 
        all.ORA.KEGG.objects[[cluster.id]] <- ora.KEGG
        kegg.ora[[cluster.id]] <- ora.KEGGdf
        }
      } else {
        kegg.ora[[cluster.id]] <- data.frame(data = c("No DE genes to conduct ORA"))
        }
    } else {
      kegg.ora[[cluster.id]] <- data.frame(data = c("No DE genes to conduct ORA"))
    }

  }
  saveRDS(all.ORA.KEGG.objects, file.path(path.to.save.output, "all_ORA_KEGG_objects.single_cell.rds"))
  saveRDS(kegg.ora, file.path(path.to.save.output, "all_ORA_KEGG_tables.single_cell.rds"))
}else {
  all.ORA.KEGG.objects <- readRDS(file.path(path.to.save.output, "all_ORA_KEGG_objects.single_cell.rds"))
  kegg.ora <- readRDS(file.path(path.to.save.output, "all_ORA_KEGG_tables.single_cell.rds"))
}

```

```{r echo=FALSE, warning=FALSE, results='asis', include=FALSE}
kegg.ora[["cluster_0"]] %>% create_dt()
```

```{r echo=FALSE, warning=FALSE, results='asis'}
for (cluster.id in names(kegg.ora)){
  cat(sprintf("#### %s \n", cluster.id))
  tmp <- kegg.ora[[cluster.id]]
  print( htmltools::tagList(datatable(tmp, extensions = 'Buttons',
                                      filter = "top",
                                      options = list(dom = 'Blfrtip',
                                                     buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                                                     lengthMenu = list(c(10,25,50,-1),
                                                                       c(10,25,50,"All")),
                                                     columnDefs = list(list(
                                                       targets = "_all",
                                                       render = JS(
                                                         "function(data, type, row, meta) {",
                                                         "return type === 'display' && data != null && data.length > 100 ?",
                                                         "'<span title=\"' + data + '\">' + data.substr(0, 100) + '...</span>' : data;",
                                                         "}")
                                                     ))
                                      ))))
  cat("\n \n")
}
```

## Gene set enrichment analysis
### GSEA-GO {.tabset}
```{r echo=FALSE, warning=FALSE, results='asis'}
#####----------------------------------------------------------------------#####
# PREPARATION
#####----------------------------------------------------------------------#####
go.gsea <- hash()
all.GSEA.GO.objects <- hash()

if (file.exists(file.path(path.to.save.output, "GO_GSEA_all_GO_objects.single_cell.rds")) == FALSE | file.exists(file.path(path.to.save.output, "GO_GSEA_pathways_tables.single_cell.rds")) == FALSE){
  for(cluster.id in all.cluster.ids){
  tmp.full.list <- de.markers.raw[[cluster.id]] %>% arrange(desc(avg_log2FC))
  input.gene.list <- tmp.full.list$avg_log2FC
  names(input.gene.list) <- tmp.full.list$Gene
  
  if (length(input.gene.list != 0)){
    GSEA.GO <- gseGO(geneList = input.gene.list,
                OrgD = org.Mm.eg.db,
                ont = "ALL",
                minGSSize = 100,
                maxGSSize = 500,
                pvalueCutoff = 0.05,
                verbose = TRUE,
                keyType = "SYMBOL", seed = TRUE)
    all.GSEA.GO.objects[[cluster.id]] <- GSEA.GO
    GSEA.GO.sum.res <- as.data.frame(GSEA.GO) 
  } else {
    GSEA.GO.sum.res <- data.frame(data = c("No DE genes to conduct ORA"))
  }
  
  GSEA.GO.sum.res <- GSEA.GO.sum.res %>% mutate_if(is.numeric, round, 6)
  go.gsea[[cluster.id]] <- GSEA.GO.sum.res  
}
  saveRDS(object = go.gsea, file = file.path(path.to.save.output, "GO_GSEA_pathways_tables.single_cell.rds"))
  saveRDS(object = all.GSEA.GO.objects, file = file.path(path.to.save.output, "GO_GSEA_all_GO_objects.single_cell.rds"))
} else {
  go.gsea <- readRDS(file.path(path.to.save.output, "GO_GSEA_pathways_tables.single_cell.rds"))
  all.GSEA.GO.objects <- readRDS(file.path(path.to.save.output, "GO_GSEA_all_GO_objects.single_cell.rds"))
}
```

```{r echo=FALSE, warning=FALSE, results='asis', include=FALSE}
diff.markers.df %>% create_dt()
```

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
for(cluster.id in all.cluster.ids){
  cat(paste("\n\n#### Cluster: ", cluster.id, "##\n"))
  tmp <- go.gsea[[cluster.id]]
  if ("core_enrichment" %in% colnames(tmp)){
    tmp <- tmp %>% subset(select = -c(core_enrichment, leading_edge, pvalue))
  }
  print( htmltools::tagList(datatable(tmp, extensions = 'Buttons',
                filter = "top",
                options = list(dom = 'Blfrtip',
                               buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                               lengthMenu = list(c(10,25,50,-1),
                                                 c(10,25,50,"All")),
                               columnDefs = list(list(
                                 targets = "_all",
                                 render = JS(
                                   "function(data, type, row, meta) {",
                                   "return type === 'display' && data != null && data.length > 100 ?",
                                   "'<span title=\"' + data + '\">' + data.substr(0, 100) + '...</span>' : data;",
                                   "}")
                               ))
                ))))
  cat("\n \n")
}
```

### GSEA-KEGG {.tabset}
```{r echo=FALSE, warning=FALSE, results='asis'}
#####----------------------------------------------------------------------#####
# PREPARATION
#####----------------------------------------------------------------------#####
kegg.gsea <- hash()
all.GSEA.KEGG.objects <- hash()

if (file.exists(file.path(path.to.save.output, "KEGG_GSEA_all_KEGG_objects.single_cell.rds")) == FALSE | file.exists(file.path(path.to.save.output, "KEGG_GSEA_pathways_tables.single_cell.rds")) == FALSE){
  for(cluster.id in all.cluster.ids){
    tmp.full.list <- de.markers.raw[[cluster.id]] %>% arrange(desc(avg_log2FC))
    convertdf <- bitr(tmp.full.list$Gene, fromType="SYMBOL", toType="ENTREZID", OrgDb="org.Mm.eg.db")
    input.gene.list <- tmp.full.list$avg_log2FC
    names(input.gene.list) <- convertdf$ENTREZID
    
    if (length(input.gene.list != 0)){
      GSEA.KEGG <- gseKEGG(geneList = input.gene.list,
                  organism = "mmu",
                  minGSSize = 100,
                  maxGSSize = 500,
                  pvalueCutoff = 0.05,
                  verbose = TRUE, seed = TRUE, keyType = "kegg")
      all.GSEA.KEGG.objects[[cluster.id]] <- GSEA.KEGG
      GSEA.KEGG.sum.res <- as.data.frame(GSEA.KEGG)
      if (nrow(GSEA.KEGG.sum.res) != 0){
        GSEA.KEGG.sum.res <- GSEA.KEGG.sum.res %>%
        rowwise() %>%
        mutate(core_enrichment = paste(to_vec(for (item in str_split(core_enrichment, "/")[[1]]) subset(convertdf, convertdf$ENTREZID == item)$SYMBOL), collapse = ", "))
      } else {
        GSEA.KEGG.sum.res <- data.frame(data = c("No enriched pathway found"))
      }
    } else {
      GSEA.KEGG.sum.res <- data.frame(data = c("No DE genes to conduct ORA"))
    }
    GSEA.KEGG.sum.res <- GSEA.KEGG.sum.res %>% mutate_if(is.numeric, round, 6) 
    kegg.gsea[[cluster.id]] <- GSEA.KEGG.sum.res
  }
  saveRDS(object = kegg.gsea, file = file.path(path.to.save.output, "KEGG_GSEA_pathways_tables.single_cell.rds"))
  saveRDS(object = all.GSEA.KEGG.objects, file = file.path(path.to.save.output, "KEGG_GSEA_all_KEGG_objects.single_cell.rds"))
} else {
  kegg.gsea <- readRDS(file.path(path.to.save.output, "KEGG_GSEA_pathways_tables.single_cell.rds"))
  all.GSEA.KEGG.objects <- readRDS(file.path(path.to.save.output, "KEGG_GSEA_all_KEGG_objects.single_cell.rds"))
}
```

```{r echo=FALSE, warning=FALSE, results='asis', include=FALSE}
diff.markers.df %>% create_dt()
```

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
for(cluster.id in all.cluster.ids){
  cat(paste("\n\n#### ", cluster.id, "##\n"))
  tmp <- kegg.gsea[[cluster.id]]
  if ("core_enrichment" %in% colnames(tmp)){
    tmp <- tmp %>% subset(select = -c(core_enrichment, leading_edge, pvalue))
  }
  print( htmltools::tagList(datatable(tmp, extensions = 'Buttons',
                filter = "top",
                options = list(dom = 'Blfrtip',
                               buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                               lengthMenu = list(c(10,25,50,-1),
                                                 c(10,25,50,"All")),
                               columnDefs = list(list(
                                 targets = "_all",
                                 render = JS(
                                   "function(data, type, row, meta) {",
                                   "return type === 'display' && data != null && data.length > 100 ?",
                                   "'<span title=\"' + data + '\">' + data.substr(0, 100) + '...</span>' : data;",
                                   "}")
                               ))
                ))))
  cat("\n \n")
}
```

# Visualization

## GO-ORA
### Dot plot depicting log10(p-value) and gene ratio {.tabset}
```{r echo=FALSE, warning=FALSE, results='asis', fig.width=15, fig.height=10}
for (cluster.id in names(all.ORA.GO.objects)){
  ora.GO <- all.ORA.GO.objects[[cluster.id]]
  if (nrow(as.data.frame(ora.GO)) == 0){
    p <- ggplot() + ggtitle("No pathway to show!")
  } else {
    p <- dotplot(ora.GO, size = "GeneRatio", showCategory=20) + ggtitle("Dotplot for Over-represnetation analysis")  
  }
  cat(sprintf("#### %s \n", cluster.id))
  print(p)
  cat("\n \n")
}
```

### Dot plot depicting log10(p-value) and background gene ratio {.tabset}
```{r echo=FALSE, warning=FALSE, results='asis', fig.width=15, fig.height=10}
for (cluster.id in names(all.ORA.GO.objects)){
  ora.GO <- all.ORA.GO.objects[[cluster.id]]
  if (nrow(as.data.frame(ora.GO)) == 0){
    p <- ggplot() + ggtitle("No pathway to show!")
  } else {
    p <- dotplot(ora.GO, size = "BgRatio", showCategory=20) + ggtitle("Dotplot for Over-represnetation analysis")  
  }
  cat(sprintf("#### %s \n", cluster.id))
  print(p)
  cat("\n \n")
}
```

## KEGG-ORA
### Dot plot depicting log10(p-value) and gene ratio {.tabset}
```{r echo=FALSE, warning=FALSE, results='asis', fig.width=15, fig.height=10}
for (cluster.id in names(all.ORA.KEGG.objects)){
  ora.KEGG <- all.ORA.KEGG.objects[[cluster.id]]
  if (nrow(as.data.frame(ora.KEGG)) == 0){
    p <- ggplot() + ggtitle("No pathway to show!")
  } else {
    p <- dotplot(ora.KEGG, size = "GeneRatio", showCategory=20) + ggtitle("Dotplot for Over-represnetation analysis")  
  }
  cat(sprintf("#### %s \n", cluster.id))
  print(p)
  cat("\n \n")
}
```

### Dot plot depicting log10(p-value) and background gene ratio {.tabset}
```{r echo=FALSE, warning=FALSE, results='asis', fig.width=15, fig.height=10}
for (cluster.id in names(all.ORA.KEGG.objects)){
  ora.KEGG <- all.ORA.KEGG.objects[[cluster.id]]
  if (nrow(as.data.frame(ora.KEGG)) == 0){
    p <- ggplot() + ggtitle("No pathway to show!")
  } else {
    p <- dotplot(ora.KEGG, size = "BgRatio", showCategory=20) + ggtitle("Dotplot for Over-represnetation analysis")  
  }
  cat(sprintf("#### %s \n", cluster.id))
  print(p)
  cat("\n \n")
}
```

## GSEA-GO {.tabset}

### Summary NES {.tabset}
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
for(cluster.id in all.cluster.ids){
  if ("data" %in% colnames(go.gsea[[cluster.id]]) == FALSE){
    cat(paste("\n\n#### Cluster: ", cluster.id, "##\n"))
    plot.df <- go.gsea[[cluster.id]] %>% subset(select = c(Description, NES)) %>% arrange(desc(NES))
    p <- ggplot(data = plot.df, aes(y = reorder(Description, NES), x = NES)) + geom_bar(stat = "identity")    
    print(p)
    cat("\n \n")
  }
}
```

## GSEA-KEGG {.tabset}
### Summary NES
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
for(cluster.id in all.cluster.ids){
  if ("data" %in% colnames(kegg.gsea[[cluster.id]]) == FALSE){
    cat(paste("\n\n#### Cluster: ", cluster.id, "##\n"))
    plot.df <- kegg.gsea[[cluster.id]] %>% subset(select = c(Description, NES)) %>% arrange(desc(NES))
    p <- ggplot(data = plot.df, aes(y = reorder(Description, NES), x = NES)) + geom_bar(stat = "identity")    
    print(p)
    cat("\n \n")
  }
}
```