---
title: "Differential gene expression analysis between sample `r params$sample1` vs. `r params$sample2`"
author:
  - "trnguyen@ukaachen.de"
date: "Last update on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    keep_md: no
    df_print: paged
    toc: true
    toc_float:
      toc_collapsed: false
    toc_depth: 3
    number_sections: true
    theme: lumen
params:
  sample1: NA
  sample2: NA
  path.to.s.obj: NA
  path.to.save.output: NA
---


```{css zoom-lib-src, echo = FALSE}
script src = "https://ajax.googleapis.com/ajax/libs/jquery/3.4.1/jquery.min.js"
```

```{js zoom-jquery, echo = FALSE}
 $(document).ready(function() {
    $('body').prepend('<div class=\"zoomDiv\"><img src=\"\" class=\"zoomImg\"></div>');
    // onClick function for all plots (img's)
    $('img:not(.zoomImg)').click(function() {
      $('.zoomImg').attr('src', $(this).attr('src')).css({width: '100%'});
      $('.zoomDiv').css({opacity: '1', width: 'auto', border: '1px solid white', borderRadius: '5px', position: 'fixed', top: '50%', left: '50%', marginRight: '-50%', transform: 'translate(-50%, -50%)', boxShadow: '0px 0px 50px #888888', zIndex: '50', overflow: 'auto', maxHeight: '100%'});
    });
    // onClick function for zoomImg
    $('img.zoomImg').click(function() {
      $('.zoomDiv').css({opacity: '0', width: '0%'}); 
    });
  });
```

<style type="text/css">
    div.datatables { height: auto !important;}
</style>


# Differential gene expression for each cluster of cells

- **positive** `logFC` indicates that the gene is highly expressed in the sample 1, `r params$sample1`.

- **negative** `logFC` indicates that the gene is highly expressed in the sample 2, `r params$sample2`.


```{r echo=FALSE, warning=FALSE, message=FALSE, results='hide', fig.width=20, fig.height=12}
#####----------------------------------------------------------------------#####
#
# trnguyen@ukaachen.de
#
#####----------------------------------------------------------------------#####

##### clean up #####
# gc()
# rm(list = ls())
if ("org.Mm.eg.db" %in% installed.packages() == FALSE){
  BiocManager::install("org.Mm.eg.db", update = FALSE)
}

if ("heatmaply" %in% installed.packages() == FALSE){
  install.packages("heatmaply")
}

library(org.Mm.eg.db)
# devtools::install_github("YuLab-SMU/yulab.utils", upgrade = "never")

#####----------------------------------------------------------------------#####
##### UPGRADE THE CLUSTERPROFILER PACKAGES
if (packageVersion("clusterProfiler") != "4.13.3"){
  remove.packages("DOSE")
  remove.packages("GOSemSim")
  remove.packages("yulab.utils")
  remove.packages("clusterProfiler")
  remotes::install_github("GuangchuangYu/GOSemSim", upgrade = "never")
  remotes::install_github("GuangchuangYu/clusterProfiler", upgrade = "never")
}
#####----------------------------------------------------------------------#####
scrna_pipeline_src <- "/home/hieunguyen/CRC1382/src_2023/src_pipeline/scRNA_GEX_pipeline_SeuratV5/processes_src"
source(file.path(scrna_pipeline_src, "import_libraries.R"))
source(file.path(scrna_pipeline_src, "helper_functions.R"))
# 
sample1 <- params$sample1
sample2 <- params$sample2
path.to.s.obj <- params$path.to.s.obj
path.to.save.output <- params$path.to.save.output

# sample1 <- "adult_SPF"
# sample2 <- "SC11"
# path.to.s.obj <- "/media/hieunguyen/CRC1382H/CRC1382/outdir/EStange_20240411_reduced_RNAcontam_0/data_analysis/03_output/remove_d4_LPS_SC5/s8_output/EStange_20240411_reduced_RNAcontam_0.output.s8.rds"
# path.to.save.output <- "/media/hieunguyen/CRC1382H/CRC1382/outdir/EStange_20240411_reduced_RNAcontam_0/data_analysis/13_output/from_03_output/remove_d4_LPS_SC5/adult_SPF_SC11/part1"

dir.create(path.to.save.output, showWarnings = FALSE, recursive = TRUE)

library(clusterProfiler)
if (packageVersion("org.Mm.eg.db") != "3.19.1"){
  install.packages("https://bioconductor.org/packages/release/bioc/src/contrib/AnnotationDbi_1.66.0.tar.gz", 
                   type = "source", repos = NULL)
  install.packages("https://bioconductor.org/packages/release/data/annotation/src/contrib/org.Mm.eg.db_3.19.1.tar.gz", 
                   type = "source", repos = NULL)
}
library(org.Mm.eg.db)
```

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
print(sprintf("Path to the input Seurat object: %s", path.to.s.obj))
print(sprintf("Sample 1: %s", sample1))
print(sprintf("Sample 2: %s", sample2))
```

```{r echo=FALSE, warning=FALSE, message=FALSE, results='hide', fig.width=20, fig.height=12}
s.obj <- readRDS(path.to.s.obj)
if ("seurat_clusters" %in% colnames(s.obj@meta.data)){
  s.obj$seurat_clusters <- NULL
}
s.obj.original <- s.obj
```

```{r echo=FALSE, warning=FALSE, message=FALSE, results='hide', fig.width=20, fig.height=12}
all.genes.in.Mm.db <- AnnotationDbi::keys(org.Mm.eg.db, keytype = "SYMBOL")
```

**In this analysis, we conduct the Differential gene expression analysis between sample** `r params$sample1` vs. `r params$sample2`  **in the dataset ** `r basename(params$path.to.s.obj)`**. We also removed TCR genes.**

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=14, fig.height=10}
TR_genes_patterns <- c("Trav", "Traj", "Trac", "Trbv", "Trbd", "Trbj", "Trbc",
                             "Trgv", "Trgj", "Trgc", "Trdv", "Trdc", "Trdj") 
DefaultAssay(s.obj) <- "SCT"

all_genes <- row.names(s.obj)
      
genes.to.exclude <- unlist(lapply(all_genes, function(x){
        if (substr(x, 1, 4) %in% TR_genes_patterns){
          return(x)
        } else {
          return(NA)
        }
}))
genes.to.exclude <- subset(genes.to.exclude, is.na(genes.to.exclude) == FALSE)
genes.to.keep <- setdiff(row.names(s.obj), genes.to.exclude)
```

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=14, fig.height=10}
s.obj <- subset(s.obj, name %in% c(sample1, sample2))
savename <- sprintf("sample_%s_vs_%s.single_cell.rds", sample1, sample2)
savename.raw <- sprintf("sample_%s_vs_%s.raw.single_cell.rds", sample1, sample2)
if (file.exists(file.path(path.to.save.output, savename)) == FALSE){
  de.markers <- hash()
  de.markers.raw <- hash()
  num.clusters <- length(unique(s.obj$cca.cluster.0.5))
  
  for (cluster.id in unique(s.obj$cca.cluster.0.5)){
    tmp.s.obj <- subset(s.obj, cca.cluster.0.5 == cluster.id)
    
    count.cells <- tmp.s.obj@meta.data %>% subset(select = c(name)) %>% table %>% data.frame()
    if (sample1 %in% count.cells$name == FALSE | sample2 %in% count.cells$name == FALSE){
      if (sample1 %in% count.cells$name == FALSE){
        count.sample1 <- 0
        count.sample2 <- subset(count.cells, count.cells$name == sample2)$Freq
      }
      if (sample2 %in% count.cells$name == FALSE){
        count.sample1 <- subset(count.cells, count.cells$name == sample1)$Freq
        count.sample2 <- 0
      }
    } else {
      count.sample1 <- subset(count.cells, count.cells$name == sample1)$Freq
      count.sample2 <- subset(count.cells, count.cells$name == sample2)$Freq
    }
    
    if (count.sample1 >= 10 & count.sample2 >= 10){
      tmp.s.obj <- PrepSCTFindMarkers(tmp.s.obj)
      tmp.de.markers <- FindMarkers(object = tmp.s.obj, 
                                    ident.1 = sample1, 
                                    ident.2 = sample2,
                                    group.by = "name", 
                                    test.use = "wilcox", 
                                    assay = "SCT",
                                    features = intersect(row.names(s.obj), genes.to.keep), 
                                    recorrect_umi = FALSE)
      
      tmp.de.markers.raw <-  tmp.de.markers %>% 
          rownames_to_column("Gene") %>%
          rowwise() %>%
          mutate(abs_avg_log2FC = abs(avg_log2FC)) %>%
          arrange(desc(abs_avg_log2FC))
      
      tmp.de.markers <- subset(tmp.de.markers, tmp.de.markers$p_val_adj <= 0.05) %>% 
          rownames_to_column("Gene") %>%
          rowwise() %>%
          mutate(abs_avg_log2FC = abs(avg_log2FC)) %>%
          arrange(desc(abs_avg_log2FC))
    } else {
      tmp.de.markers <- data.frame(data = c("Not enough cells to conduct DE test, either sample1 <= 10 cells or sample2 <= 10 cells"))
      tmp.de.markers.raw <- data.frame(data = c("Not enough cells to conduct DE test, either sample1 <= 10 cells or sample2 <= 10 cells"))
    }
    de.markers[[sprintf("cluster_%s", cluster.id)]] <- tmp.de.markers
    de.markers.raw[[sprintf("cluster_%s", cluster.id)]] <- tmp.de.markers.raw
  }
  saveRDS(object = de.markers, file.path(path.to.save.output, savename))
  saveRDS(object = de.markers.raw, file.path(path.to.save.output, savename.raw))
} else {
  de.markers <- readRDS(file.path(path.to.save.output, savename))
  de.markers.raw <- readRDS(file.path(path.to.save.output, savename.raw))
}
```

## Full tables of all D.E genes between sample `r params$sample1` vs `r params$sample2`` {.tabset}
```{r echo=FALSE, include=FALSE}
de.markers[[sprintf("cluster_%s", 0)]] %>% create_dt()
```

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
all.clusters.in.de.markers <- sort(to_vec( for (item in names(de.markers)) as.numeric(str_split(item, "_")[[1]][[2]]) ))
all.clusters.in.de.markers <- to_vec( for(item in all.clusters.in.de.markers) sprintf("cluster_%s", item))
for (cluster.id in all.clusters.in.de.markers){
  tmp <- de.markers[[cluster.id]] %>% mutate_if(is.numeric, round, 5) 
  cat(paste("\n\n###", cluster.id, "##\n"))
  print( htmltools::tagList(datatable(tmp, extensions = 'Buttons',
                filter = "top",
                options = list(dom = 'Blfrtip',
                               buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                               lengthMenu = list(c(10,25,50,-1),
                                                 c(10,25,50,"All")),
                               columnDefs = list(list(
                                 targets = "_all",
                                 render = JS(
                                   "function(data, type, row, meta) {",
                                   "return type === 'display' && data != null && data.length > 100 ?",
                                   "'<span title=\"' + data + '\">' + data.substr(0, 100) + '...</span>' : data;",
                                   "}")
                               ))
                ))))
  cat("\n \n")
}
```

## Volcano plot {.tabset}

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=14, fig.height=10}
for (cluster.id in all.clusters.in.de.markers){
  cat(sprintf("### Cluster %s \n", cluster.id))
  input.df <- de.markers.raw[[cluster.id]] 
  if ("data" %in% colnames(input.df) == FALSE & nrow(input.df) != 0){
    top10up.genes <- input.df %>% subset(p_val_adj <= 0.05) %>% arrange(desc(avg_log2FC)) %>% head(10)
    top10down.genes <- input.df %>% subset(p_val_adj <= 0.05) %>% arrange(desc(avg_log2FC)) %>% tail(10)
    input.df <- input.df %>%
    rowwise() %>%
    mutate(sig = ifelse(p_val_adj <= 0.05, "Sig.", "not Sig.")) %>%
      mutate(show.gene.name = ifelse(Gene %in% c(top10up.genes$Gene, top10down.genes$Gene), Gene, NA))
      
    
    cutoff.adjp <- 0.05
    volcano.plot <- ggplot(data=input.df, 
                           aes(x=avg_log2FC, y=-log10(p_val_adj), col = sig, label=Gene)) + 
      geom_point(size = 3) + geom_label_repel(label = input.df$show.gene.name, size = 8) + 
      scale_color_manual(values=c("#c0d2f0", "#f28095")) +
      geom_vline(xintercept=c(-1, 1), col="#9a9fa6") +
      geom_hline(yintercept=-log10(cutoff.adjp), col="#9a9fa6") +
      theme(plot.title = element_text(hjust=0.5, face="bold", size = 12), axis.text=element_text(size=12)) + 
      theme_bw()
    
  } else {
    volcano.plot <- ggplot() + ggtitle("Not enough cells to conduct DE test")
  }
  print(volcano.plot)
  cat("\n \n")
}

```

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
library("clusterProfiler")
```

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
diff.markers.df <- data.frame()
for (cluster.id in names(de.markers.raw)){
  tmp <- de.markers.raw[[cluster.id]] %>% mutate(cluster = cluster.id)
  if ("data" %in% colnames(tmp) == FALSE){
    diff.markers.df <- rbind(diff.markers.df, tmp)    
  }
}
if ("org.Mm.eg.db" %in% installed.packages() == FALSE){
  BiocManager::install("org.Mm.eg.db")
}
library("org.Mm.eg.db")
```

## Heatmaps {.tabset}
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=15}
mapal <- colorRampPalette(RColorBrewer::brewer.pal(11,"RdBu"))(256)
for (cluster.id in all.clusters.in.de.markers){
  cat(sprintf("### Cluster %s \n", cluster.id))
  inputdf <- de.markers[[cluster.id]] 
  if ("data" %in% colnames(inputdf) | nrow(inputdf) == 0){
    p <- ggplot() + ggtitle("No significantly DE genes found or not enough cells to conduct DE test!")
  } else {
    top10up.genes <- inputdf %>% subset(p_val_adj <= 0.05 & avg_log2FC > 0) %>% arrange(desc(avg_log2FC)) %>% head(20)
    top10down.genes <- inputdf %>% subset(p_val_adj <= 0.05 & avg_log2FC < 0) %>% arrange(desc(avg_log2FC)) %>% tail(20)
    
    if (length(c(top10up.genes$Gene, top10down.genes$Gene)) > 0){
      p <- DoHeatmap(object = subset(s.obj, seurat_clusters == as.numeric(str_replace(cluster.id, "cluster_", ""))), 
                     features = c(top10up.genes$Gene, top10down.genes$Gene), group.by = "name", assay = "RNA") + 
        scale_fill_gradientn(colours = rev(mapal))
    } else {
      p <- ggplot() + ggtitle("No significantly DE genes found or not enough cells to conduct DE test!")
    }
  }
  print(p)
  cat("\n \n")
}
```

# Pathway analysis for the cluster specific DE test (single cell data)
## Full list of genes ranked by `avg_log2FC` {.tabset}

```{r echo=FALSE, warning=FALSE, results='asis', include=FALSE}
diff.markers.df <- data.frame()
for (cluster.id in names(de.markers.raw)){
  tmp <- de.markers.raw[[cluster.id]] %>% mutate(cluster = cluster.id)
  if ("data" %in% colnames(tmp) == FALSE){
    diff.markers.df <- rbind(diff.markers.df, tmp)    
  }
}
```

```{r echo=FALSE, warning=FALSE, results='asis', include=FALSE}
diff.markers.df %>% create_dt()
```

```{r echo=FALSE, warning=FALSE, results='asis'}
cat("List of significantly differentially expressed genes between 2 samples \n")

all.cluster.ids <- unique(diff.markers.df$cluster)

all.cluster.ids.idx <- sort(to_vec(for (item in all.cluster.ids)
  as.numeric(str_replace(item, "cluster_", ""))))
all.cluster.ids <- to_vec(for (item in all.cluster.ids.idx)
  sprintf("cluster_%s", item))

for(cluster.id in all.cluster.ids){
  df.cluster.id <- subset(diff.markers.df, diff.markers.df$cluster == cluster.id) %>% mutate_if(is.numeric, round, 6)
  cat(paste("\n\n### Cluster: ", cluster.id, "##\n"))
  print( htmltools::tagList(datatable(df.cluster.id, extensions = 'Buttons',
                filter = "top",
                options = list(dom = 'Blfrtip',
                               buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                               lengthMenu = list(c(10,25,50,-1),
                                                 c(10,25,50,"All")),
                               columnDefs = list(list(
                                 targets = "_all",
                                 render = JS(
                                   "function(data, type, row, meta) {",
                                   "return type === 'display' && data != null && data.length > 100 ?",
                                   "'<span title=\"' + data + '\">' + data.substr(0, 100) + '...</span>' : data;",
                                   "}")
                               ))
                ))))
  cat("\n \n")
}

```

## Genes of interest (DE genes only) {.tabset } 
```{r echo=FALSE, warning=FALSE, results='asis', include=FALSE}
diff.markers.df %>% create_dt()
```

```{r echo=FALSE, warning=FALSE, results='asis'}

for(cluster.id in all.cluster.ids){
  df.cluster.id <- subset(diff.markers.df, diff.markers.df$cluster == cluster.id & diff.markers.df$p_val_adj <= 0.05) %>% mutate_if(is.numeric, round, 6)
  cat(paste("\n\n### Cluster: ", cluster.id, "##\n"))
    print( htmltools::tagList(datatable(df.cluster.id, extensions = 'Buttons',
                filter = "top",
                options = list(dom = 'Blfrtip',
                               buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                               lengthMenu = list(c(10,25,50,-1),
                                                 c(10,25,50,"All")),
                               columnDefs = list(list(
                                 targets = "_all",
                                 render = JS(
                                   "function(data, type, row, meta) {",
                                   "return type === 'display' && data != null && data.length > 100 ?",
                                   "'<span title=\"' + data + '\">' + data.substr(0, 100) + '...</span>' : data;",
                                   "}")
                               ))
                ))))
  cat("\n \n")
}
```

## Over-representation analyiss (ORA)

### GO-ORA {.tabset}
```{r echo=FALSE, warning=FALSE, results='asis'}
if (file.exists(file.path(path.to.save.output, "all_ORA_GO_objects.single_cell.rds")) == FALSE){
  go.ora <- hash()
  all.ORA.GO.objects <- hash()
  for (cluster.id in all.cluster.ids){
    input.gene.list <- de.markers[[cluster.id]]$Gene
    if (length(input.gene.list) != 0){
      ora.GO <- enrichGO(gene = input.gene.list,
                        OrgDb = org.Mm.eg.db,
                        ont = "ALL",
                        pvalueCutoff = 0.05,
                        qvalueCutoff = 0.05,
                        readable = TRUE,
                        keyType = "SYMBOL",
                        pAdjustMethod = "BH")
      
      ora.GOdf <- as.data.frame(ora.GO)
      if (nrow(ora.GOdf) == 0){
        go.ora[[cluster.id]] <- data.frame(data = c("No DE genes to conduct ORA  or no pathway found"))
      } else {
        ora.GOdf <- ora.GOdf[order(ora.GOdf$p.adjust, decreasing = FALSE),]  %>% mutate_if(is.numeric, round, 6) 
        all.ORA.GO.objects[[cluster.id]] <- ora.GO
        go.ora[[cluster.id]] <- ora.GOdf
      }
    } else {
      go.ora[[cluster.id]] <- data.frame(data = c("No DE genes to conduct ORA"))
    }
  }
  saveRDS(all.ORA.GO.objects, file.path(path.to.save.output, "all_ORA_GO_objects.single_cell.rds"))
  saveRDS(go.ora, file.path(path.to.save.output, "all_ORA_GO_tables.single_cell.rds"))
} else {
  all.ORA.GO.objects <- readRDS(file.path(path.to.save.output, "all_ORA_GO_objects.single_cell.rds"))
  go.ora <- readRDS(file.path(path.to.save.output, "all_ORA_GO_tables.single_cell.rds"))
}
```

```{r echo=FALSE, warning=FALSE, results='asis', include=FALSE}
go.ora[["cluster_0"]] %>% create_dt()
```

```{r echo=FALSE, warning=FALSE, results='asis'}
for (cluster.id in names(go.ora)){
  cat(sprintf("#### %s \n", cluster.id))
  tmp <- go.ora[[cluster.id]]
  print( htmltools::tagList(datatable(tmp, extensions = 'Buttons',
                filter = "top",
                options = list(dom = 'Blfrtip',
                               buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                               lengthMenu = list(c(10,25,50,-1),
                                                 c(10,25,50,"All")),
                               columnDefs = list(list(
                                 targets = "_all",
                                 render = JS(
                                   "function(data, type, row, meta) {",
                                   "return type === 'display' && data != null && data.length > 100 ?",
                                   "'<span title=\"' + data + '\">' + data.substr(0, 100) + '...</span>' : data;",
                                   "}")
                               ))
                ))))
  cat("\n \n")
}
```

### KEGG-ORA {.tabset}
```{r echo=FALSE, warning=FALSE, results='asis', message=FALSE}
if (file.exists(file.path(path.to.save.output, "all_ORA_KEGG_objects.single_cell.rds")) == FALSE){
  kegg.ora <- hash()
  all.ORA.KEGG.objects <- hash()
  for (cluster.id in all.cluster.ids){
    input.gene.list <- de.markers[[cluster.id]]$Gene 
    if (length(input.gene.list) != 0) {
      if (length(intersect(input.gene.list, all.genes.in.Mm.db)) != 0){
        input.gene.list <- bitr(input.gene.list, fromType="SYMBOL", toType="ENTREZID", OrgDb="org.Mm.eg.db")
        ora.KEGG <-  enrichKEGG(gene = input.gene.list$ENTREZID,
                         organism     = 'mmu',
                         pvalueCutoff = 0.05, 
                         qvalueCutoff = 0.05)
        ora.KEGGdf <- as.data.frame(ora.KEGG)
        if (nrow(ora.KEGGdf) == 0){
          kegg.ora[[cluster.id]] <- data.frame(data = c("No DE genes to conduct ORA  or no pathway found"))
        } else {
                ora.KEGGdf <- ora.KEGGdf[order(ora.KEGGdf$p.adjust), ] %>% mutate_if(is.numeric, round, 6) 
        all.ORA.KEGG.objects[[cluster.id]] <- ora.KEGG
        kegg.ora[[cluster.id]] <- ora.KEGGdf
        }
      } else {
        kegg.ora[[cluster.id]] <- data.frame(data = c("No DE genes to conduct ORA"))
        }
    } else {
      kegg.ora[[cluster.id]] <- data.frame(data = c("No DE genes to conduct ORA"))
    }

  }
  saveRDS(all.ORA.KEGG.objects, file.path(path.to.save.output, "all_ORA_KEGG_objects.single_cell.rds"))
  saveRDS(kegg.ora, file.path(path.to.save.output, "all_ORA_KEGG_tables.single_cell.rds"))
}else {
  all.ORA.KEGG.objects <- readRDS(file.path(path.to.save.output, "all_ORA_KEGG_objects.single_cell.rds"))
  kegg.ora <- readRDS(file.path(path.to.save.output, "all_ORA_KEGG_tables.single_cell.rds"))
}

```

```{r echo=FALSE, warning=FALSE, results='asis', include=FALSE}
kegg.ora[["cluster_0"]] %>% create_dt()
```

```{r echo=FALSE, warning=FALSE, results='asis'}
for (cluster.id in names(kegg.ora)){
  cat(sprintf("#### %s \n", cluster.id))
  tmp <- kegg.ora[[cluster.id]]
  print( htmltools::tagList(datatable(tmp, extensions = 'Buttons',
                                      filter = "top",
                                      options = list(dom = 'Blfrtip',
                                                     buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                                                     lengthMenu = list(c(10,25,50,-1),
                                                                       c(10,25,50,"All")),
                                                     columnDefs = list(list(
                                                       targets = "_all",
                                                       render = JS(
                                                         "function(data, type, row, meta) {",
                                                         "return type === 'display' && data != null && data.length > 100 ?",
                                                         "'<span title=\"' + data + '\">' + data.substr(0, 100) + '...</span>' : data;",
                                                         "}")
                                                     ))
                                      ))))
  cat("\n \n")
}
```

## Gene set enrichment analysis
### GSEA-GO {.tabset}
```{r echo=FALSE, warning=FALSE, results='asis'}
#####----------------------------------------------------------------------#####
# PREPARATION
#####----------------------------------------------------------------------#####
go.gsea <- hash()
all.GSEA.GO.objects <- hash()

if (file.exists(file.path(path.to.save.output, "GO_GSEA_all_GO_objects.single_cell.rds")) == FALSE | file.exists(file.path(path.to.save.output, "GO_GSEA_pathways_tables.single_cell.rds")) == FALSE){
  for(cluster.id in all.cluster.ids){
  tmp.full.list <- de.markers.raw[[cluster.id]] %>% arrange(desc(avg_log2FC))
  input.gene.list <- tmp.full.list$avg_log2FC
  names(input.gene.list) <- tmp.full.list$Gene
  
  if (length(input.gene.list != 0)){
    GSEA.GO <- gseGO(geneList = input.gene.list,
                OrgD = org.Mm.eg.db,
                ont = "ALL",
                minGSSize = 100,
                maxGSSize = 500,
                pvalueCutoff = 0.05,
                verbose = TRUE,
                keyType = "SYMBOL", seed = TRUE)
    all.GSEA.GO.objects[[cluster.id]] <- GSEA.GO
    GSEA.GO.sum.res <- as.data.frame(GSEA.GO) 
  } else {
    GSEA.GO.sum.res <- data.frame(data = c("No DE genes to conduct ORA"))
  }
  
  GSEA.GO.sum.res <- GSEA.GO.sum.res %>% mutate_if(is.numeric, round, 6)
  go.gsea[[cluster.id]] <- GSEA.GO.sum.res  
}
  saveRDS(object = go.gsea, file = file.path(path.to.save.output, "GO_GSEA_pathways_tables.single_cell.rds"))
  saveRDS(object = all.GSEA.GO.objects, file = file.path(path.to.save.output, "GO_GSEA_all_GO_objects.single_cell.rds"))
} else {
  go.gsea <- readRDS(file.path(path.to.save.output, "GO_GSEA_pathways_tables.single_cell.rds"))
  all.GSEA.GO.objects <- readRDS(file.path(path.to.save.output, "GO_GSEA_all_GO_objects.single_cell.rds"))
}
```

```{r echo=FALSE, warning=FALSE, results='asis', include=FALSE}
diff.markers.df %>% create_dt()
```

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
for(cluster.id in all.cluster.ids){
  cat(paste("\n\n#### Cluster: ", cluster.id, "##\n"))
  tmp <- go.gsea[[cluster.id]]
  if ("core_enrichment" %in% colnames(tmp)){
    tmp <- tmp %>% subset(select = -c(core_enrichment, leading_edge, pvalue))
  }
  print( htmltools::tagList(datatable(tmp, extensions = 'Buttons',
                filter = "top",
                options = list(dom = 'Blfrtip',
                               buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                               lengthMenu = list(c(10,25,50,-1),
                                                 c(10,25,50,"All")),
                               columnDefs = list(list(
                                 targets = "_all",
                                 render = JS(
                                   "function(data, type, row, meta) {",
                                   "return type === 'display' && data != null && data.length > 100 ?",
                                   "'<span title=\"' + data + '\">' + data.substr(0, 100) + '...</span>' : data;",
                                   "}")
                               ))
                ))))
  cat("\n \n")
}
```

### GSEA-KEGG {.tabset}
```{r echo=FALSE, warning=FALSE, results='asis'}
#####----------------------------------------------------------------------#####
# PREPARATION
#####----------------------------------------------------------------------#####
kegg.gsea <- hash()
all.GSEA.KEGG.objects <- hash()


if (file.exists(file.path(path.to.save.output, "KEGG_GSEA_all_KEGG_objects.single_cell.rds")) == FALSE | file.exists(file.path(path.to.save.output, "KEGG_GSEA_pathways_tables.single_cell.rds")) == FALSE){
  for(cluster.id in all.cluster.ids){
    tmp.full.list <- de.markers.raw[[cluster.id]] %>% arrange(desc(avg_log2FC))
    convertdf <- bitr(tmp.full.list$Gene, fromType="SYMBOL", toType="ENTREZID", OrgDb="org.Mm.eg.db")
    input.gene.list <- tmp.full.list$avg_log2FC
    names(input.gene.list) <- convertdf$ENTREZID
    
    if (length(input.gene.list != 0)){
      GSEA.KEGG <- gseKEGG(geneList = input.gene.list,
                  organism = "mmu",
                  minGSSize = 100,
                  maxGSSize = 500,
                  pvalueCutoff = 0.05,
                  verbose = TRUE, seed = TRUE, keyType = "kegg")
      all.GSEA.KEGG.objects[[cluster.id]] <- GSEA.KEGG
      GSEA.KEGG.sum.res <- as.data.frame(GSEA.KEGG)
      if (nrow(GSEA.KEGG.sum.res) != 0){
        GSEA.KEGG.sum.res <- GSEA.KEGG.sum.res %>%
        rowwise() %>%
        mutate(core_enrichment = paste(to_vec(for (item in str_split(core_enrichment, "/")[[1]]) subset(convertdf, convertdf$ENTREZID == item)$SYMBOL), collapse = ", "))
      } else {
        GSEA.KEGG.sum.res <- data.frame(data = c("No enriched pathway found"))
      }
    } else {
      GSEA.KEGG.sum.res <- data.frame(data = c("No DE genes to conduct ORA"))
    }
    
    GSEA.KEGG.sum.res <- GSEA.KEGG.sum.res %>% mutate_if(is.numeric, round, 6) 
    kegg.gsea[[cluster.id]] <- GSEA.KEGG.sum.res
  }
  saveRDS(object = kegg.gsea, file = file.path(path.to.save.output, "KEGG_GSEA_pathways_tables.single_cell.rds"))
  saveRDS(object = all.GSEA.KEGG.objects, file = file.path(path.to.save.output, "KEGG_GSEA_all_KEGG_objects.single_cell.rds"))
} else {
  kegg.gsea <- readRDS(file.path(path.to.save.output, "KEGG_GSEA_pathways_tables.single_cell.rds"))
  all.GSEA.KEGG.objects <- readRDS(file.path(path.to.save.output, "KEGG_GSEA_all_KEGG_objects.single_cell.rds"))
}

```

```{r echo=FALSE, warning=FALSE, results='asis', include=FALSE}
diff.markers.df %>% create_dt()
```

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
for(cluster.id in all.cluster.ids){
  cat(paste("\n\n#### ", cluster.id, "##\n"))
  tmp <- kegg.gsea[[cluster.id]]
  if ("core_enrichment" %in% colnames(tmp)){
    tmp <- tmp %>% subset(select = -c(core_enrichment, leading_edge, pvalue))
  }
  print( htmltools::tagList(datatable(tmp, extensions = 'Buttons',
                filter = "top",
                options = list(dom = 'Blfrtip',
                               buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                               lengthMenu = list(c(10,25,50,-1),
                                                 c(10,25,50,"All")),
                               columnDefs = list(list(
                                 targets = "_all",
                                 render = JS(
                                   "function(data, type, row, meta) {",
                                   "return type === 'display' && data != null && data.length > 100 ?",
                                   "'<span title=\"' + data + '\">' + data.substr(0, 100) + '...</span>' : data;",
                                   "}")
                               ))
                ))))
  cat("\n \n")
}
```

# Visualization

## GO-ORA
### Dot plot depicting log10(p-value) and gene ratio {.tabset}
```{r echo=FALSE, warning=FALSE, results='asis', fig.width=15, fig.height=10}
for (cluster.id in names(all.ORA.GO.objects)){
  ora.GO <- all.ORA.GO.objects[[cluster.id]]
  if (nrow(as.data.frame(ora.GO)) == 0){
    p <- ggplot() + ggtitle("No pathway to show!")
  } else {
    p <- dotplot(ora.GO, size = "GeneRatio", showCategory=20) + ggtitle("Dotplot for Over-represnetation analysis")  
  }
  cat(sprintf("#### %s \n", cluster.id))
  print(p)
  cat("\n \n")
}


```

### Dot plot depicting log10(p-value) and background gene ratio {.tabset}
```{r echo=FALSE, warning=FALSE, results='asis', fig.width=15, fig.height=10}
for (cluster.id in names(all.ORA.GO.objects)){
  ora.GO <- all.ORA.GO.objects[[cluster.id]]
  if (nrow(as.data.frame(ora.GO)) == 0){
    p <- ggplot() + ggtitle("No pathway to show!")
  } else {
    p <- dotplot(ora.GO, size = "BgRatio", showCategory=20) + ggtitle("Dotplot for Over-represnetation analysis")  
  }
  cat(sprintf("#### %s \n", cluster.id))
  print(p)
  cat("\n \n")
}


```

## KEGG-ORA
### Dot plot depicting log10(p-value) and gene ratio {.tabset}
```{r echo=FALSE, warning=FALSE, results='asis', fig.width=15, fig.height=10}
for (cluster.id in names(all.ORA.KEGG.objects)){
  ora.KEGG <- all.ORA.KEGG.objects[[cluster.id]]
  if (nrow(as.data.frame(ora.KEGG)) == 0){
    p <- ggplot() + ggtitle("No pathway to show!")
  } else {
    p <- dotplot(ora.KEGG, size = "GeneRatio", showCategory=20) + ggtitle("Dotplot for Over-represnetation analysis")  
  }
  cat(sprintf("#### %s \n", cluster.id))
  print(p)
  cat("\n \n")
}
```

### Dot plot depicting log10(p-value) and background gene ratio {.tabset}
```{r echo=FALSE, warning=FALSE, results='asis', fig.width=15, fig.height=10}
for (cluster.id in names(all.ORA.KEGG.objects)){
  ora.KEGG <- all.ORA.KEGG.objects[[cluster.id]]
  if (nrow(as.data.frame(ora.KEGG)) == 0){
    p <- ggplot() + ggtitle("No pathway to show!")
  } else {
    p <- dotplot(ora.KEGG, size = "BgRatio", showCategory=20) + ggtitle("Dotplot for Over-represnetation analysis")  
  }
  cat(sprintf("#### %s \n", cluster.id))
  print(p)
  cat("\n \n")
}


```



## GSEA-GO {.tabset}

### Summary NES {.tabset}
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
for(cluster.id in all.cluster.ids){
  if ("data" %in% colnames(go.gsea[[cluster.id]]) == FALSE){
    cat(paste("\n\n#### Cluster: ", cluster.id, "##\n"))
    plot.df <- go.gsea[[cluster.id]] %>% subset(select = c(Description, NES)) %>% arrange(desc(NES))
    p <- ggplot(data = plot.df, aes(y = reorder(Description, NES), x = NES)) + geom_bar(stat = "identity")    
    print(p)
    cat("\n \n")
  }
}

```


## GSEA-KEGG {.tabset}

### Summary NES
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
for(cluster.id in all.cluster.ids){
  if ("data" %in% colnames(kegg.gsea[[cluster.id]]) == FALSE){
    cat(paste("\n\n#### Cluster: ", cluster.id, "##\n"))
    plot.df <- kegg.gsea[[cluster.id]] %>% subset(select = c(Description, NES)) %>% arrange(desc(NES))
    p <- ggplot(data = plot.df, aes(y = reorder(Description, NES), x = NES)) + geom_bar(stat = "identity")    
    print(p)
    cat("\n \n")
  }
}
```

# Pseudobulk differential gene expression test for all cells

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=14, fig.height=10}
library(DESeq2)
s.obj.full <- s.obj.original

s.obj <- hash()
s.obj[[sample1]] <- subset(s.obj.full, name == sample1)
s.obj[[sample2]] <- subset(s.obj.full, name == sample2)
```

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=14, fig.height=10}
DimPlot(object = subset(s.obj.full, name %in% c(sample1, sample2)), reduction = "cca_UMAP", label = TRUE, split.by = "name", ncol = 2)
```


```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
DefaultAssay(s.obj[[sample1]]) <- "SCT"
DefaultAssay(s.obj[[sample2]]) <- "SCT"

hashtag.cells <- hash()
hashtag.cells[[sample1]] <- hash()
hashtag.cells[[sample2]] <- hash()

all.ht <- hash()

all.ht[[sample1]] <- setdiff(unique(s.obj[[sample1]]$HTO_classification), c("Negative"))
all.ht[[sample2]] <- setdiff(unique(s.obj[[sample2]]$HTO_classification), c("Negative"))

for (sample.id in c(sample1, sample2)){
  true.exists.ht <- c()
  for (ht in all.ht[[sample.id]]){
    ht.cells <- colnames(subset(s.obj[[sample.id]], HTO_classification == ht))
    if (length(ht.cells) > 1){
      hashtag.cells[[sample.id]][[ht]] <- ht.cells      
      true.exists.ht <- c(true.exists.ht, ht)
    }
  }
  all.ht[[sample.id]] <- true.exists.ht
}

count.matrix <- hash()

pseudobulk.matrix <- data.frame(Gene = rownames(s.obj[[sample1]]))
for (sample.id in c(sample1, sample2)){
  count.matrix[[sample.id]] <- hash()
  tmp.count.matrix <- GetAssayData(object = s.obj[[sample.id]], slot = "counts", assay = "SCT")
  for (ht in all.ht[[sample.id]]){
    tmp <- tmp.count.matrix[, hashtag.cells[[sample.id]][[ht]]]
    rowsum.tmp <- rowSums(tmp)
    tmpdf <- data.frame(data = rowsum.tmp[rowsum.tmp >= 10]) %>% rownames_to_column("Gene")
    colnames(tmpdf) <- c("Gene", sprintf("%s_%s", sample.id, ht))
    count.matrix[[sample.id]][[ht]] <- tmpdf
    pseudobulk.matrix <- merge(pseudobulk.matrix, tmpdf, by.x = "Gene", by.y = "Gene")
  }
}

pseudobulk.matrix <- pseudobulk.matrix %>% column_to_rownames("Gene")
pseudo_metadata <- data.frame(sample.id = colnames(pseudobulk.matrix))
pseudo_metadata$group <- unlist(lapply(pseudo_metadata$sample.id, function(x){
  str_split(x, "_Hashtag")[[1]][[1]]
}))

pseudo_metadata$group <- factor(pseudo_metadata$group, levels = c(sample1, sample2))

dds <- DESeqDataSetFromMatrix(pseudobulk.matrix, 
                              colData = pseudo_metadata, 
                              design = ~ group)

dds <- DESeq(dds)
contrast <- c("group", levels(pseudo_metadata$group)[2], levels(pseudo_metadata$group)[1])

# resultsNames(ddsrow
res <- results(dds, 
               contrast = contrast,
               alpha = 0.05)

resdf <- res %>%
  data.frame() %>% rownames_to_column("Gene")

padj_cutoff <- 0.05
logFC_cutoff <- 1

normalized_counts <- counts(dds, normalized = TRUE) %>% as.data.frame() %>% rownames_to_column("Gene")
resdf <- merge(resdf, normalized_counts, by.x = "Gene", by.y = "Gene") %>%
  mutate(abslog2FoldChange = abs(log2FoldChange)) %>%
  mutate(sig = ifelse(padj <= padj_cutoff & abslog2FoldChange >= logFC_cutoff, "sig.diff", "nonsig.diff")) 

resdf.sigOnly <- subset(resdf, resdf$padj <= padj_cutoff & resdf$abslog2FoldChange >= logFC_cutoff) %>% arrange(desc(log2FoldChange))
heatmap.input <- rbind(head(resdf.sigOnly, 10), tail(resdf.sigOnly, 10))[, c("Gene", pseudo_metadata$sample.id)]

heatmap.top50.input <- rbind(head(resdf.sigOnly, 50), tail(resdf.sigOnly, 50))[, c("Gene", pseudo_metadata$sample.id)]

heatmap.full.input <-resdf.sigOnly[, c("Gene", pseudo_metadata$sample.id)] %>% column_to_rownames("Gene")
heatmap.full.input.scaled <- (heatmap.full.input - rowSums(heatmap.full.input))/rowSds(as.matrix(heatmap.full.input)) 

heatmap.full.input.scaled <- heatmap.full.input.scaled %>% rownames_to_column("Gene") 
heatmap.full.input.scaled <- merge(heatmap.full.input.scaled, subset(resdf.sigOnly, select = c(Gene, padj, log2FoldChange, abslog2FoldChange)),
                                   by.x = "Gene", by.y = "Gene") %>% arrange(desc(log2FoldChange))
```

## Metadata (pseudobulk, 1 hashtag = 1 sample)
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=28, fig.height=20}
pseudo_metadata %>% create_dt()
```

## Table: list of all significantly differential expressed genes
- **Positive** `logFC` indicates that the gene is up in sample 1, `r params$sample1`.

- **Negative** `logFC` indicates that the gene is up in sample 2, `r params$sample2`.

This table is sorted by the column `log2FoldChange` in descending order. 

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=28, fig.height=20}
resdf.sigOnly %>% mutate_if(is.numeric, round, 8) %>% create_dt()
```

## Scaled z-score pseudobulk gene expression, all DE genes
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
writexl::write_xlsx(heatmap.full.input.scaled, file.path(path.to.save.output, sprintf("heatmap_input_%s_%s_allDEGs.pseudobulk.scaled.xlsx", sample1, sample2)))
```


```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
print("Saved output file, not displaying here")
# heatmap.full.input.scaled %>% create_dt()
```

## Heatmap
Heatmap for top10 up/down regulated genes in the comparison `r params$sample1` vs `r params$sample2``.

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=15}
library(heatmaply)
if (nrow(heatmap.input) == 0 ){
  ggplot() + ggtitle("No significant genes to show!")
} else {
  heatmaply(log10(heatmap.input[, pseudo_metadata$sample.id]),
          main= sprintf("Heatmap, Sample: %s vs. %s ", sample1, sample2),
          method = "plotly",labRow=heatmap.input$Gene,
          xlab = "Samples", ylab = "Genes", width = 800, height = 600,
          showticklabels = c(TRUE, TRUE), show_dendrogram = c(FALSE, TRUE),
          key.title = "log10 scale colormap",
          label_names = c("Gene", "Sample", "Expression"),
          k_col = 2)
   # heatmaply(log10(heatmap.top50.input[, pseudo_metadata$sample.id]),
   #        main= sprintf("Heatmap, Sample: %s vs. %s ", sample1, sample2),
   #        method = "plotly",labRow=heatmap.top50.input$Gene,
   #        xlab = "Samples", ylab = "Genes", width = 800, height = 600,
   #        showticklabels = c(TRUE, TRUE), show_dendrogram = c(FALSE, TRUE),
   #        key.title = "log10 scale colormap",
   #        label_names = c("Gene", "Sample", "Expression"),
   #        k_col = 2) 
}


```

## Volcano plot
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=15, fig.height=10}
volcano.plot <- ggplot(data=resdf, 
                       aes(x=log2FoldChange, y=-log10(padj), col=sig, label=Gene)) + 
  geom_point() + 
  scale_color_manual(values=c("#c0d2f0", "#f28095")) +
  theme_minimal() +
  geom_vline(xintercept=c(-1, 1), col="#9a9fa6", linetype='dotted') +
  geom_hline(yintercept=-log10(padj_cutoff), col="#9a9fa6", linetype='dotted') +
  ggtitle(sprintf("Volcano plot, Sample: %s vs. %s", sample1, sample2)) +
  theme_bw() + 
  theme(plot.title = element_text(hjust=0.5, face="bold", size = 12))
ggplotly(volcano.plot)
```

# Pathway analysis with GO Terms and KEGG database, general information


## Over-representation analysis (ORA)

- `ONTOLOGY`: BP, MF, and CC represent Biological Process, Molecular Function, and Cellular Component groups of GO. In this analysis, we use all these three sub-ontologies.

- There are two p-values calculated by the ORA: `pvalue` and `p.adjust`, we will use only the `p.adjust` and `q.value` to decide if a pathway is enriched or not. Method of False discovery rate: Benjamin - Hochberg. Thresholds are: `p.value <= 0.05` and `p.adjust <= 0.05` and `q.value <= 0.05`. 

- Explanation on the output of ORA summary table:

  - `GeneRatio`: `GeneRatio = k/n`, where `k` is the number of genes within your input gene list (genes of interest) which are annotated to the gene set and `n` is the size of the list of genes of interest. 
  
  - `BgRatio`: `BgRatio = M/N`, where `M` is the number of genes within the hypergeometric distribution (the distribution we used to calculate the p-value) that are annotated (either directly or indirectly) to the genes of interest. `N` is the total number of genes in the background distribution (by default, it is all the genes that have annotation). 
  

- In the ORA, our gene set of interest is the **significantly differentially expressed genes** found from previous analysis. 

## Gene set enrichment analysis (GSEA)

- All genes will be used in the GSEA. Genes are ranked based on their average log2 Fold-change, which was found from the DE analysis. 

- GSEA aggregates the per gene statistics across genes within a gene set, therefore making it possible to detect situations where all genes in a predefined set change in a small but coordinated way.

- Given apriori defined set of gene S, the goal of GSEA is to determine whether the members of S are randomly distributed throughout the ranked gene list (L) or primarily found at the top or bottom.

- There are three key elements of the GSEA method:

  - Calculation of an Enrichment Score.
  
  - Estimation of Significance Level of ES.
  
  - Adjustment for Multiple Hypothesis Testing.

For all significant pathways, we generate a bar-plot showing all Normalized Enrichment Scores.


**CLICK ON THE IMAGE TO ZOOM-IN AND ONE MORE CLICK TO GO BACK TO THE ORIGINAL**

**If the table is empty, that means; no significant pathways were found**

**Updated note:**

`set.seed(42)` and set `seed = TRUE` in the functions of `clusterProfiler` to ensure reproducibility.


# Pathway analysis results for the pseudobulk SCT DE test

```{r, echo=FALSE, message=FALSE, results='asis', fig.width=14, fig.height=10}
diff.markers.df <- resdf

```

## Full list of genes ranked by `avg_log2FC` {.tabset}

Here we show all genes in the dataset, including both significantly DE genes and non-significantly DE genes. Genes are ranked in descending order of the log2 Fold-change. This list of gene is used in the GENE SET ENRICHMENT ANALYSIS. 

```{r echo=FALSE, warning=FALSE, results='asis', }
diff.markers.df %>% mutate_if(is.numeric, round, 6) %>% arrange(desc(log2FoldChange)) %>% create_dt()
```

## Significantly Differentially expressed genes only
For OVER REPRESENTATION ANALYSIS, we use only significantly DE genes. 
```{r echo=FALSE, warning=FALSE, results='asis'}
sig.diff.markers.df <- subset(diff.markers.df, diff.markers.df$sig == "sig.diff")
sig.diff.markers.df %>% mutate_if(is.numeric, round, 6) %>% arrange(desc(log2FoldChange)) %>% create_dt()
```

## Over-representation analysis (ORA)


### GO-ORA
```{r echo=FALSE, warning=FALSE, results='asis'}
input.gene.list <- sig.diff.markers.df$Gene
ora.GO <- enrichGO(gene = input.gene.list,
                    OrgDb = org.Mm.eg.db,
                    ont = "ALL",
                    pvalueCutoff = 0.05,
                    qvalueCutoff = 0.05,
                    readable = TRUE,
                    keyType = "SYMBOL",
                    pAdjustMethod = "BH")

if (is.null(ora.GO) == TRUE){
  ora.GOdf <- data.frame(status = c("No results from ORA - GO"))  
} else {
  ora.GOdf <- as.data.frame(ora.GO)
  ora.GOdf <- ora.GOdf[order(ora.GOdf$p.adjust, decreasing = FALSE),]  
}

ora.GOdf %>% mutate_if(is.numeric, round, 6) %>% create_dt()
```

### KEGG-ORA
```{r echo=FALSE, warning=FALSE, results='asis', message=FALSE}
input.gene.list <- subset(diff.markers.df, diff.markers.df$padj <= 0.05)$Gene 

if (length(input.gene.list) <= 1 | length(intersect(all.genes.in.Mm.db, input.gene.list)) == 0){
  ora.KEGGdf <- data.frame(status = c("No gene or only 1 DE gene available"))
} else {
  input.gene.list <- bitr(input.gene.list, fromType="SYMBOL", toType="ENTREZID", OrgDb="org.Mm.eg.db")  
  ora.KEGG <-  enrichKEGG(gene = input.gene.list$ENTREZID,
                 organism     = 'mmu',
                 pvalueCutoff = 0.05, 
                 qvalueCutoff = 0.05)
  if (is.null(ora.KEGG) == TRUE){
    ora.KEGGdf <- data.frame(status = c("No results obtained from KEGG-ORA"))
  } else {
    ora.KEGGdf <- as.data.frame(ora.KEGG)
    ora.KEGGdf <- ora.KEGGdf[order(ora.KEGGdf$p.adjust), ]    
  }
}

ora.KEGGdf %>% mutate_if(is.numeric, round, 6) %>% create_dt()
```

## Gene set enrichment analysis (GSEA)
### GSEA-GO
```{r echo=FALSE, warning=FALSE, results='asis'}
tmp.full.list <- diff.markers.df %>% arrange(desc(log2FoldChange))
input.gene.list <- tmp.full.list$log2FoldChange
names(input.gene.list) <- tmp.full.list$Gene

GSEA.GO <- gseGO(geneList = input.gene.list,
              OrgD = org.Mm.eg.db,
              ont = "ALL",
              minGSSize = 100,
              maxGSSize = 500,
              pvalueCutoff = 0.05,
              verbose = TRUE,
              keyType = "SYMBOL", seed = TRUE)

GSEA.GOdf <- as.data.frame(GSEA.GO) 

GSEA.GOdf <- GSEA.GOdf %>% rowwise() %>% 
  mutate(abs.NES = abs(NES)) %>%
  rownames_to_column("idx")

GSEA.GOdf <- GSEA.GOdf[order(GSEA.GOdf$NES, decreasing = TRUE), ]

top10_up_nes_pw.GO <- head(GSEA.GOdf, 10)$Description
top10_down_nes_pw.GO <- tail(GSEA.GOdf, 10)$Description

top10_up_nes_idx.GO <- head(GSEA.GOdf, 10)$idx
top10_down_nes_idx.GO <- tail(GSEA.GOdf, 10)$idx

GSEA.GOdf %>% mutate_if(is.numeric, round, 6) %>% create_dt()

```

### GSEA-KEGG
```{r echo=FALSE, warning=FALSE, results='asis'}
tmp.full.list <- diff.markers.df %>% arrange(desc(log2FoldChange))

convertdf <- bitr(tmp.full.list$Gene, fromType="SYMBOL", toType="ENTREZID", OrgDb="org.Mm.eg.db")
input.gene.list <- tmp.full.list$log2FoldChange
names(input.gene.list) <- convertdf$ENTREZID

GSEA.KEGG <- gseKEGG(geneList = input.gene.list,
              organism = "mmu",
              minGSSize = 100,
              maxGSSize = 500,
              pvalueCutoff = 0.05,
              verbose = FALSE, seed = TRUE)

GSEA.KEGGdf <- as.data.frame(GSEA.KEGG)

GSEA.KEGGdf <- GSEA.KEGGdf %>% rowwise() %>% 
  mutate(abs.NES = abs(NES)) %>%
  rownames_to_column("idx")

GSEA.KEGGdf <- GSEA.KEGGdf[order(GSEA.KEGGdf$NES, decreasing = TRUE), ]

top10_up_nes_pw.KEGG <- head(GSEA.KEGGdf, 10)$Description
top10_down_nes_pw.KEGG <- tail(GSEA.KEGGdf, 10)$Description

top10_up_nes_idx.KEGG <- head(GSEA.KEGGdf, 10)$idx
top10_down_nes_idx.KEGG <- tail(GSEA.KEGGdf, 10)$idx

GSEA.KEGGdf %>% mutate_if(is.numeric, round, 6) %>% create_dt()
```

# Visualization for the pathway analysis

## GO-ORA
### Dot plot depicting log10(p-value) and gene ratio
```{r echo=FALSE, warning=FALSE, results='asis', fig.width=15, fig.height=10}
if (nrow(ora.GOdf) == 0 | "status" %in% colnames(ora.GOdf)){
  ggplot() + ggtitle("No pathway to show!")
} else {
  dotplot(ora.GO, size = "GeneRatio", showCategory=20) + ggtitle("Dotplot for Over-represnetation analysis")  
}

```

### Dot plot depicting log10(p-value) and background gene ratio
```{r echo=FALSE, warning=FALSE, results='asis', fig.width=15, fig.height=10}
if (nrow(ora.GOdf) == 0 | "status" %in% colnames(ora.GOdf)){
  ggplot() + ggtitle("No pathway to show!")
} else {
  dotplot(ora.GO, size = "BgRatio", showCategory=20) + ggtitle("Dotplot for Over-represnetation analysis")
}
```

## KEGG-ORA

### Dot plot depicting log10(p-value) and gene ratio
```{r echo=FALSE, warning=FALSE, results='asis', fig.width=15, fig.height=10}
if (nrow(ora.KEGGdf) == 0 | "status" %in% colnames(ora.KEGGdf)){
  ggplot() + ggtitle("No pathway to show!")
} else {
  dotplot(ora.KEGG, size = "GeneRatio", showCategory=20) + ggtitle("Dotplot for Over-represnetation analysis")
}
```

### Dot plot depicting log10(p-value) and background gene ratio
```{r echo=FALSE, warning=FALSE, results='asis', fig.width=15, fig.height=10}
if (nrow(ora.KEGGdf) == 0 | "status" %in% colnames(ora.KEGGdf)){
  ggplot() + ggtitle("No pathway to show!")
} else {
  dotplot(ora.KEGG, size = "BgRatio", showCategory=20) + ggtitle("Dotplot for Over-represnetation analysis")
}
```

## GO-GSEA
### Summary pathways by NES values
```{r echo=FALSE, warning=FALSE, results='asis', fig.width=20, fig.height=10}
plot.df <- GSEA.GOdf %>% 
  subset(Description %in% c(top10_up_nes_pw.GO, top10_down_nes_pw.GO)) %>% 
  subset(select = c(Description, NES)) %>% arrange(desc(NES))

p <- ggplot(data = plot.df, aes(y = reorder(Description, NES), x = NES)) + geom_bar(stat = "identity")    

print(p)
```

### Top-10 up NES {.tabset}
```{r echo=FALSE, warning=FALSE, results='asis', fig.width=15, fig.height=10}
for (i in top10_up_nes_idx.GO){
  i <- as.numeric(i)
  p1 <- gseaplot(GSEA.GO, geneSetID = i, by = "runningScore", title = GSEA.GO$Description[i])
  p2 <- gseaplot(GSEA.GO, geneSetID = i, by = "preranked", title = GSEA.GO$Description[i])
  p3 <- gseaplot(GSEA.GO, geneSetID = i, title = GSEA.GO$Description[1])
  p <- cowplot::plot_grid(p1, p2, p3, ncol=1, labels=LETTERS[1:3])
  pw <- GSEA.GO$Description[i]
  cat(sprintf("#### %s \n", pw))
  print(p)
  cat("\n \n")
}

```


### Top-10 down NES {.tabset}
```{r echo=FALSE, warning=FALSE, results='asis', fig.width=15, fig.height=10}
for (i in top10_down_nes_idx.GO){
  i <- as.numeric(i)
  p1 <- gseaplot(GSEA.GO, geneSetID = i, by = "runningScore", title = GSEA.GO$Description[i])
  p2 <- gseaplot(GSEA.GO, geneSetID = i, by = "preranked", title = GSEA.GO$Description[i])
  p3 <- gseaplot(GSEA.GO, geneSetID = i, title = GSEA.GO$Description[1])
  p <- cowplot::plot_grid(p1, p2, p3, ncol=1, labels=LETTERS[1:3])
  pw <- GSEA.GO$Description[i]
  cat(sprintf("#### %s \n", pw))
  print(p)
  cat("\n \n")
}

```

## KEGG-GSEA
### Summary pathways by NES values
```{r echo=FALSE, warning=FALSE, results='asis', fig.width=20, fig.height=10}
plot.df <- GSEA.KEGGdf  %>% 
  subset(Description %in% c(top10_up_nes_pw.KEGG, top10_down_nes_pw.KEGG)) %>% 
  subset(select = c(Description, NES)) %>% arrange(desc(NES))

p <- ggplot(data = plot.df, aes(y = reorder(Description, NES), x = NES)) + geom_bar(stat = "identity")    

print(p)

```

### Top-10 up NES {.tabset}
```{r echo=FALSE, warning=FALSE, results='asis', fig.width=15, fig.height=10}
for (i in top10_up_nes_idx.KEGG){
  i <- as.numeric(i)
  p1 <- gseaplot(GSEA.KEGG, geneSetID = i, by = "runningScore", title = GSEA.KEGG$Description[i])
  p2 <- gseaplot(GSEA.KEGG, geneSetID = i, by = "preranked", title = GSEA.KEGG$Description[i])
  p3 <- gseaplot(GSEA.KEGG, geneSetID = i, title = GSEA.KEGG$Description[1])
  p <- cowplot::plot_grid(p1, p2, p3, ncol=1, labels=LETTERS[1:3])
  pw <- GSEA.KEGG$Description[i]
  cat(sprintf("#### %s \n", pw))
  print(p)
  cat("\n \n")
}
```

### Top-10 down NES {.tabset}
```{r echo=FALSE, warning=FALSE, results='asis', fig.width=15, fig.height=10}
for (i in top10_down_nes_idx.KEGG){
  i <- as.numeric(i)
  p1 <- gseaplot(GSEA.KEGG, geneSetID = i, by = "runningScore", title = GSEA.KEGG$Description[i])
  p2 <- gseaplot(GSEA.KEGG, geneSetID = i, by = "preranked", title = GSEA.KEGG$Description[i])
  p3 <- gseaplot(GSEA.KEGG, geneSetID = i, title = GSEA.KEGG$Description[1])
  p <- cowplot::plot_grid(p1, p2, p3, ncol=1, labels=LETTERS[1:3])
  pw <- GSEA.KEGG$Description[i]
  cat(sprintf("#### %s \n", pw))
  print(p)
  cat("\n \n")
}

```

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=28, fig.height=20}
```

