---
title: "Differential gene expression analysis between sample `r params$sample1` vs. `r params$sample2`, pseudobulk DGE for each cluster"
author:
  - "trnguyen@ukaachen.de"
date: "Last update on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    keep_md: no
    df_print: paged
    toc: true
    toc_float:
      toc_collapsed: false
    toc_depth: 3
    number_sections: true
    theme: lumen
params:
  sample1: NA
  sample2: NA
  path.to.s.obj: NA
  path.to.save.output: NA
  cluster.id: NA
---

```{css zoom-lib-src, echo = FALSE}
script src = "https://ajax.googleapis.com/ajax/libs/jquery/3.4.1/jquery.min.js"
```

```{js zoom-jquery, echo = FALSE}
 $(document).ready(function() {
    $('body').prepend('<div class=\"zoomDiv\"><img src=\"\" class=\"zoomImg\"></div>');
    // onClick function for all plots (img's)
    $('img:not(.zoomImg)').click(function() {
      $('.zoomImg').attr('src', $(this).attr('src')).css({width: '100%'});
      $('.zoomDiv').css({opacity: '1', width: 'auto', border: '1px solid white', borderRadius: '5px', position: 'fixed', top: '50%', left: '50%', marginRight: '-50%', transform: 'translate(-50%, -50%)', boxShadow: '0px 0px 50px #888888', zIndex: '50', overflow: 'auto', maxHeight: '100%'});
    });
    // onClick function for zoomImg
    $('img.zoomImg').click(function() {
      $('.zoomDiv').css({opacity: '0', width: '0%'}); 
    });
  });
```

<style type="text/css">
    div.datatables { height: auto !important;}
</style>


# Differential gene expression for each cluster of cells

- **positive** `logFC` indicates that the gene is highly expressed in the sample 1, `r params$sample1`.

- **negative** `logFC` indicates that the gene is highly expressed in the sample 2, `r params$sample2`.


```{r echo=FALSE, warning=FALSE, message=FALSE, results='hide', fig.width=20, fig.height=12}
#####----------------------------------------------------------------------#####
#
# trnguyen@ukaachen.de
#
#####----------------------------------------------------------------------#####

##### clean up #####
# gc()
# rm(list = ls())
if ("org.Mm.eg.db" %in% installed.packages() == FALSE){
  BiocManager::install("org.Mm.eg.db", update = FALSE)
}

if ("heatmaply" %in% installed.packages() == FALSE){
  install.packages("heatmaply")
}

library(org.Mm.eg.db)

#####----------------------------------------------------------------------#####
##### UPGRADE THE CLUSTERPROFILER PACKAGES
#####----------------------------------------------------------------------#####
scrna_pipeline_src <- "/home/hieunguyen/CRC1382/src_2023/src_pipeline/scRNA_GEX_pipeline_SeuratV5/processes_src"
source(file.path(scrna_pipeline_src, "import_libraries.R"))
source(file.path(scrna_pipeline_src, "helper_functions.R"))

sample1 <- params$sample1
sample2 <- params$sample2
cluster.id <- params$cluster.id
path.to.s.obj <- params$path.to.s.obj
path.to.save.output <- params$path.to.save.output
cluster.id <- params$cluster.id

# sample1 <- "adult_SPF"
# sample2 <- "SC11"
# cluster.id <- 21
# path.to.s.obj <- "/media/hieunguyen/CRC1382H/CRC1382/outdir/EStange_20240411_reduced_RNAcontam_0/data_analysis/03_output/remove_d4_LPS_SC5/s8_output/EStange_20240411_reduced_RNAcontam_0.output.s8.rds"
# path.to.save.output <- "/media/hieunguyen/CRC1382H/CRC1382/outdir/EStange_20240411_reduced_RNAcontam_0/data_analysis/13_output/from_03_output/remove_d4_LPS_SC5/adult_SPF_SC11/part2/cluster_21"

dir.create(path.to.save.output, showWarnings = FALSE, recursive = TRUE)

library(clusterProfiler)
if (packageVersion("org.Mm.eg.db") != "3.19.1"){
  install.packages("https://bioconductor.org/packages/release/bioc/src/contrib/AnnotationDbi_1.66.0.tar.gz", 
                   type = "source", repos = NULL)
  install.packages("https://bioconductor.org/packages/release/data/annotation/src/contrib/org.Mm.eg.db_3.19.1.tar.gz", 
                   type = "source", repos = NULL)
}
library(org.Mm.eg.db)
```

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
print(sprintf("Path to the input Seurat object: %s", path.to.s.obj))
print(sprintf("Sample 1: %s", sample1))
print(sprintf("Sample 2: %s", sample2))
```

```{r echo=FALSE, warning=FALSE, message=FALSE, results='hide', fig.width=20, fig.height=12}
s.obj.original <- readRDS(path.to.s.obj)
if ("seurat_clusters" %in% colnames(s.obj.original@meta.data)){
  s.obj.original$seurat_clusters <- NULL
}

```

```{r echo=FALSE, warning=FALSE, message=FALSE, results='hide', fig.width=20, fig.height=12}
all.genes.in.Mm.db <- AnnotationDbi::keys(org.Mm.eg.db, keytype = "SYMBOL")
```

**In this analysis, we conduct the Differential gene expression analysis between sample** `r params$sample1` vs. `r params$sample2`  **in the dataset ** `r basename(params$path.to.s.obj)`**. We also removed TCR genes.**

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=14, fig.height=10}
TR_genes_patterns <- c("Trav", "Traj", "Trac", "Trbv", "Trbd", "Trbj", "Trbc",
                             "Trgv", "Trgj", "Trgc", "Trdv", "Trdc", "Trdj") 
DefaultAssay(s.obj.original) <- "SCT"

all_genes <- row.names(s.obj.original)
      
genes.to.exclude <- unlist(lapply(all_genes, function(x){
        if (substr(x, 1, 4) %in% TR_genes_patterns){
          return(x)
        } else {
          return(NA)
        }
}))
genes.to.exclude <- subset(genes.to.exclude, is.na(genes.to.exclude) == FALSE)
genes.to.keep <- setdiff(row.names(s.obj.original), genes.to.exclude)
```

# Pseudobulk differential gene expression

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=14, fig.height=10}
library(DESeq2)
input.s.obj <- s.obj.original

path.to.project.src <- "/home/hieunguyen/CRC1382/src_2023/SBharadwaj/deep_seq_batch"
source(file.path(path.to.project.src, "08_helper_functions.R"))
```

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
all.clusters <- unique(s.obj.original$cca.cluster.0.5)

input.s.obj <- subset(s.obj.original, cca.cluster.0.5 == cluster.id)
pseudobulk.dge.output <- run_pseudobulk_dge(input.s.obj = input.s.obj, sample1 = sample1, sample2 = sample2)

dds = pseudobulk.dge.output$dds
resdf = pseudobulk.dge.output$resdf
resdf.sigOnly = pseudobulk.dge.output$resdf.sigOnly
heatmap.top50.input = pseudobulk.dge.output$heatmap.top50.input
heatmap.full.input.scaled = pseudobulk.dge.output$heatmap.full.input.scaled
pseudo_metadata = pseudobulk.dge.output$pseudo_metadata

heatmap.input <- rbind(head(resdf.sigOnly, 10), tail(resdf.sigOnly, 10))[, c("Gene", pseudo_metadata$sample.id)]
padj_cutoff <- 0.05
logFC_cutoff <- 1
```

## Metadata (pseudobulk, 1 hashtag = 1 sample)
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=28, fig.height=20}
pseudo_metadata %>% create_dt()
```

## Table: list of all significantly differential expressed genes
- **Positive** `logFC` indicates that the gene is up in sample 1, `r params$sample1`.

- **Negative** `logFC` indicates that the gene is up in sample 2, `r params$sample2`.

This table is sorted by the column `log2FoldChange` in descending order. 

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=28, fig.height=20}
resdf.sigOnly %>% mutate_if(is.numeric, round, 8) %>% create_dt()
```

## Scaled z-score pseudobulk gene expression, all DE genes
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
writexl::write_xlsx(heatmap.full.input.scaled, file.path(path.to.save.output, sprintf("heatmap_input_%s_%s_allDEGs.pseudobulk.scaled.xlsx", sample1, sample2)))
```


```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=12}
print("Saved output file, not displaying here")
# heatmap.full.input.scaled %>% create_dt()
```

## Heatmap
Heatmap for top10 up/down regulated genes in the comparison `r params$sample1` vs `r params$sample2``.

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=20, fig.height=15}
library(heatmaply)
if (nrow(heatmap.input) == 0 ){
  ggplot() + ggtitle("No significant genes to show!")
} else {
  heatmaply(log10(heatmap.input[, pseudo_metadata$sample.id]),
          main= sprintf("Heatmap, Sample: %s vs. %s ", sample1, sample2),
          method = "plotly",labRow=heatmap.input$Gene,
          xlab = "Samples", ylab = "Genes", width = 800, height = 600,
          showticklabels = c(TRUE, TRUE), show_dendrogram = c(FALSE, TRUE),
          key.title = "log10 scale colormap",
          label_names = c("Gene", "Sample", "Expression"),
          k_col = 2)
   # heatmaply(log10(heatmap.top50.input[, pseudo_metadata$sample.id]),
   #        main= sprintf("Heatmap, Sample: %s vs. %s ", sample1, sample2),
   #        method = "plotly",labRow=heatmap.top50.input$Gene,
   #        xlab = "Samples", ylab = "Genes", width = 800, height = 600,
   #        showticklabels = c(TRUE, TRUE), show_dendrogram = c(FALSE, TRUE),
   #        key.title = "log10 scale colormap",
   #        label_names = c("Gene", "Sample", "Expression"),
   #        k_col = 2) 
}


```

## Volcano plot
```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=15, fig.height=10}
volcano.plot <- ggplot(data=resdf, 
                       aes(x=log2FoldChange, y=-log10(padj), col=sig, label=Gene)) + 
  geom_point() + 
  scale_color_manual(values=c("#c0d2f0", "#f28095")) +
  theme_minimal() +
  geom_vline(xintercept=c(-1, 1), col="#9a9fa6", linetype='dotted') +
  geom_hline(yintercept=-log10(padj_cutoff), col="#9a9fa6", linetype='dotted') +
  ggtitle(sprintf("Volcano plot, Sample: %s vs. %s", sample1, sample2)) +
  theme_bw() + 
  theme(plot.title = element_text(hjust=0.5, face="bold", size = 12))
ggplotly(volcano.plot)
```

# Pathway analysis with GO Terms and KEGG database, general information


## Over-representation analysis (ORA)

- `ONTOLOGY`: BP, MF, and CC represent Biological Process, Molecular Function, and Cellular Component groups of GO. In this analysis, we use all these three sub-ontologies.

- There are two p-values calculated by the ORA: `pvalue` and `p.adjust`, we will use only the `p.adjust` and `q.value` to decide if a pathway is enriched or not. Method of False discovery rate: Benjamin - Hochberg. Thresholds are: `p.value <= 0.05` and `p.adjust <= 0.05` and `q.value <= 0.05`. 

- Explanation on the output of ORA summary table:

  - `GeneRatio`: `GeneRatio = k/n`, where `k` is the number of genes within your input gene list (genes of interest) which are annotated to the gene set and `n` is the size of the list of genes of interest. 
  
  - `BgRatio`: `BgRatio = M/N`, where `M` is the number of genes within the hypergeometric distribution (the distribution we used to calculate the p-value) that are annotated (either directly or indirectly) to the genes of interest. `N` is the total number of genes in the background distribution (by default, it is all the genes that have annotation). 
  

- In the ORA, our gene set of interest is the **significantly differentially expressed genes** found from previous analysis. 

## Gene set enrichment analysis (GSEA)

- All genes will be used in the GSEA. Genes are ranked based on their average log2 Fold-change, which was found from the DE analysis. 

- GSEA aggregates the per gene statistics across genes within a gene set, therefore making it possible to detect situations where all genes in a predefined set change in a small but coordinated way.

- Given apriori defined set of gene S, the goal of GSEA is to determine whether the members of S are randomly distributed throughout the ranked gene list (L) or primarily found at the top or bottom.

- There are three key elements of the GSEA method:

  - Calculation of an Enrichment Score.
  
  - Estimation of Significance Level of ES.
  
  - Adjustment for Multiple Hypothesis Testing.

For all significant pathways, we generate a bar-plot showing all Normalized Enrichment Scores.


**CLICK ON THE IMAGE TO ZOOM-IN AND ONE MORE CLICK TO GO BACK TO THE ORIGINAL**

**If the table is empty, that means; no significant pathways were found**

**Updated note:**

`set.seed(42)` and set `seed = TRUE` in the functions of `clusterProfiler` to ensure reproducibility.


# Pathway analysis results for the pseudobulk SCT DE test

```{r, echo=FALSE, message=FALSE, results='asis', fig.width=14, fig.height=10}
diff.markers.df <- resdf

```

## Full list of genes ranked by `avg_log2FC` {.tabset}

Here we show all genes in the dataset, including both significantly DE genes and non-significantly DE genes. Genes are ranked in descending order of the log2 Fold-change. This list of gene is used in the GENE SET ENRICHMENT ANALYSIS. 

```{r echo=FALSE, warning=FALSE, results='asis', }
diff.markers.df %>% mutate_if(is.numeric, round, 6) %>% arrange(desc(log2FoldChange)) %>% create_dt()
```

## Significantly Differentially expressed genes only
For OVER REPRESENTATION ANALYSIS, we use only significantly DE genes. 
```{r echo=FALSE, warning=FALSE, results='asis'}
sig.diff.markers.df <- subset(diff.markers.df, diff.markers.df$sig == "sig.diff")
sig.diff.markers.df %>% mutate_if(is.numeric, round, 6) %>% arrange(desc(log2FoldChange)) %>% create_dt()
```

## Over-representation analysis (ORA)


### GO-ORA
```{r echo=FALSE, warning=FALSE, results='asis'}
input.gene.list <- sig.diff.markers.df$Gene
ora.GO <- enrichGO(gene = input.gene.list,
                    OrgDb = org.Mm.eg.db,
                    ont = "ALL",
                    pvalueCutoff = 0.05,
                    qvalueCutoff = 0.05,
                    readable = TRUE,
                    keyType = "SYMBOL",
                    pAdjustMethod = "BH")

if (is.null(ora.GO) == TRUE){
  ora.GOdf <- data.frame(status = c("No results from ORA - GO"))  
} else {
  ora.GOdf <- as.data.frame(ora.GO)
  ora.GOdf <- ora.GOdf[order(ora.GOdf$p.adjust, decreasing = FALSE),]  
}

ora.GOdf %>% mutate_if(is.numeric, round, 6) %>% create_dt()
```

### KEGG-ORA
```{r echo=FALSE, warning=FALSE, results='asis', message=FALSE}
input.gene.list <- subset(diff.markers.df, diff.markers.df$padj <= 0.05)$Gene 

if (length(input.gene.list) <= 1 | length(intersect(all.genes.in.Mm.db, input.gene.list)) == 0){
  ora.KEGGdf <- data.frame(status = c("No gene or only 1 DE gene available"))
} else {
  input.gene.list <- bitr(input.gene.list, fromType="SYMBOL", toType="ENTREZID", OrgDb="org.Mm.eg.db")  
  ora.KEGG <-  enrichKEGG(gene = input.gene.list$ENTREZID,
                 organism     = 'mmu',
                 pvalueCutoff = 0.05, 
                 qvalueCutoff = 0.05)
  if (is.null(ora.KEGG) == TRUE){
    ora.KEGGdf <- data.frame(status = c("No results obtained from KEGG-ORA"))
  } else {
    ora.KEGGdf <- as.data.frame(ora.KEGG)
    ora.KEGGdf <- ora.KEGGdf[order(ora.KEGGdf$p.adjust), ]    
  }
}

ora.KEGGdf %>% mutate_if(is.numeric, round, 6) %>% create_dt()
```

## Gene set enrichment analysis (GSEA)
### GSEA-GO
```{r echo=FALSE, warning=FALSE, results='asis'}
tmp.full.list <- diff.markers.df %>% arrange(desc(log2FoldChange))
input.gene.list <- tmp.full.list$log2FoldChange
names(input.gene.list) <- tmp.full.list$Gene

GSEA.GO <- gseGO(geneList = input.gene.list,
              OrgD = org.Mm.eg.db,
              ont = "ALL",
              minGSSize = 100,
              maxGSSize = 500,
              pvalueCutoff = 0.05,
              verbose = TRUE,
              keyType = "SYMBOL", seed = TRUE)

GSEA.GOdf <- as.data.frame(GSEA.GO) 

GSEA.GOdf <- GSEA.GOdf %>% rowwise() %>% 
  mutate(abs.NES = abs(NES)) %>%
  rownames_to_column("idx")

GSEA.GOdf <- GSEA.GOdf[order(GSEA.GOdf$NES, decreasing = TRUE), ]

top10_up_nes_pw.GO <- head(GSEA.GOdf, 10)$Description
top10_down_nes_pw.GO <- tail(GSEA.GOdf, 10)$Description

top10_up_nes_idx.GO <- head(GSEA.GOdf, 10)$idx
top10_down_nes_idx.GO <- tail(GSEA.GOdf, 10)$idx

GSEA.GOdf %>% mutate_if(is.numeric, round, 6) %>% create_dt()

```

### GSEA-KEGG
```{r echo=FALSE, warning=FALSE, results='asis'}
tmp.full.list <- diff.markers.df %>% arrange(desc(log2FoldChange))

convertdf <- bitr(tmp.full.list$Gene, fromType="SYMBOL", toType="ENTREZID", OrgDb="org.Mm.eg.db")
input.gene.list <- tmp.full.list$log2FoldChange
names(input.gene.list) <- convertdf$ENTREZID

GSEA.KEGG <- gseKEGG(geneList = input.gene.list,
              organism = "mmu",
              minGSSize = 100,
              maxGSSize = 500,
              pvalueCutoff = 0.05,
              verbose = FALSE, seed = TRUE)

GSEA.KEGGdf <- as.data.frame(GSEA.KEGG)

GSEA.KEGGdf <- GSEA.KEGGdf %>% rowwise() %>% 
  mutate(abs.NES = abs(NES)) %>%
  rownames_to_column("idx")

GSEA.KEGGdf <- GSEA.KEGGdf[order(GSEA.KEGGdf$NES, decreasing = TRUE), ]

top10_up_nes_pw.KEGG <- head(GSEA.KEGGdf, 10)$Description
top10_down_nes_pw.KEGG <- tail(GSEA.KEGGdf, 10)$Description

top10_up_nes_idx.KEGG <- head(GSEA.KEGGdf, 10)$idx
top10_down_nes_idx.KEGG <- tail(GSEA.KEGGdf, 10)$idx

GSEA.KEGGdf %>% mutate_if(is.numeric, round, 6) %>% create_dt()
```

# Visualization for the pathway analysis

## GO-ORA
### Dot plot depicting log10(p-value) and gene ratio
```{r echo=FALSE, warning=FALSE, results='asis', fig.width=15, fig.height=10}
if (nrow(ora.GOdf) == 0 | "status" %in% colnames(ora.GOdf)){
  ggplot() + ggtitle("No pathway to show!")
} else {
  dotplot(ora.GO, size = "GeneRatio", showCategory=20) + ggtitle("Dotplot for Over-represnetation analysis")  
}

```

### Dot plot depicting log10(p-value) and background gene ratio
```{r echo=FALSE, warning=FALSE, results='asis', fig.width=15, fig.height=10}
if (nrow(ora.GOdf) == 0 | "status" %in% colnames(ora.GOdf)){
  ggplot() + ggtitle("No pathway to show!")
} else {
  dotplot(ora.GO, size = "BgRatio", showCategory=20) + ggtitle("Dotplot for Over-represnetation analysis")
}
```

## KEGG-ORA

### Dot plot depicting log10(p-value) and gene ratio
```{r echo=FALSE, warning=FALSE, results='asis', fig.width=15, fig.height=10}
if (nrow(ora.KEGGdf) == 0 | "status" %in% colnames(ora.KEGGdf)){
  ggplot() + ggtitle("No pathway to show!")
} else {
  dotplot(ora.KEGG, size = "GeneRatio", showCategory=20) + ggtitle("Dotplot for Over-represnetation analysis")
}
```

### Dot plot depicting log10(p-value) and background gene ratio
```{r echo=FALSE, warning=FALSE, results='asis', fig.width=15, fig.height=10}
if (nrow(ora.KEGGdf) == 0 | "status" %in% colnames(ora.KEGGdf)){
  ggplot() + ggtitle("No pathway to show!")
} else {
  dotplot(ora.KEGG, size = "BgRatio", showCategory=20) + ggtitle("Dotplot for Over-represnetation analysis")
}
```

## GO-GSEA
### Summary pathways by NES values
```{r echo=FALSE, warning=FALSE, results='asis', fig.width=20, fig.height=10}
plot.df <- GSEA.GOdf %>% 
  subset(Description %in% c(top10_up_nes_pw.GO, top10_down_nes_pw.GO)) %>% 
  subset(select = c(Description, NES)) %>% arrange(desc(NES))

p <- ggplot(data = plot.df, aes(y = reorder(Description, NES), x = NES)) + geom_bar(stat = "identity")    

print(p)
```

### Top-10 up NES {.tabset}
```{r echo=FALSE, warning=FALSE, results='asis', fig.width=15, fig.height=10}
for (i in top10_up_nes_idx.GO){
  i <- as.numeric(i)
  p1 <- gseaplot(GSEA.GO, geneSetID = i, by = "runningScore", title = GSEA.GO$Description[i])
  p2 <- gseaplot(GSEA.GO, geneSetID = i, by = "preranked", title = GSEA.GO$Description[i])
  p3 <- gseaplot(GSEA.GO, geneSetID = i, title = GSEA.GO$Description[1])
  p <- cowplot::plot_grid(p1, p2, p3, ncol=1, labels=LETTERS[1:3])
  pw <- GSEA.GO$Description[i]
  cat(sprintf("#### %s \n", pw))
  print(p)
  cat("\n \n")
}

```


### Top-10 down NES {.tabset}
```{r echo=FALSE, warning=FALSE, results='asis', fig.width=15, fig.height=10}
for (i in top10_down_nes_idx.GO){
  i <- as.numeric(i)
  p1 <- gseaplot(GSEA.GO, geneSetID = i, by = "runningScore", title = GSEA.GO$Description[i])
  p2 <- gseaplot(GSEA.GO, geneSetID = i, by = "preranked", title = GSEA.GO$Description[i])
  p3 <- gseaplot(GSEA.GO, geneSetID = i, title = GSEA.GO$Description[1])
  p <- cowplot::plot_grid(p1, p2, p3, ncol=1, labels=LETTERS[1:3])
  pw <- GSEA.GO$Description[i]
  cat(sprintf("#### %s \n", pw))
  print(p)
  cat("\n \n")
}

```

## KEGG-GSEA
### Summary pathways by NES values
```{r echo=FALSE, warning=FALSE, results='asis', fig.width=20, fig.height=10}
plot.df <- GSEA.KEGGdf  %>% 
  subset(Description %in% c(top10_up_nes_pw.KEGG, top10_down_nes_pw.KEGG)) %>% 
  subset(select = c(Description, NES)) %>% arrange(desc(NES))

p <- ggplot(data = plot.df, aes(y = reorder(Description, NES), x = NES)) + geom_bar(stat = "identity")    

print(p)

```

### Top-10 up NES {.tabset}
```{r echo=FALSE, warning=FALSE, results='asis', fig.width=15, fig.height=10}
for (i in top10_up_nes_idx.KEGG){
  i <- as.numeric(i)
  p1 <- gseaplot(GSEA.KEGG, geneSetID = i, by = "runningScore", title = GSEA.KEGG$Description[i])
  p2 <- gseaplot(GSEA.KEGG, geneSetID = i, by = "preranked", title = GSEA.KEGG$Description[i])
  p3 <- gseaplot(GSEA.KEGG, geneSetID = i, title = GSEA.KEGG$Description[1])
  p <- cowplot::plot_grid(p1, p2, p3, ncol=1, labels=LETTERS[1:3])
  pw <- GSEA.KEGG$Description[i]
  cat(sprintf("#### %s \n", pw))
  print(p)
  cat("\n \n")
}
```

### Top-10 down NES {.tabset}
```{r echo=FALSE, warning=FALSE, results='asis', fig.width=15, fig.height=10}
for (i in top10_down_nes_idx.KEGG){
  i <- as.numeric(i)
  p1 <- gseaplot(GSEA.KEGG, geneSetID = i, by = "runningScore", title = GSEA.KEGG$Description[i])
  p2 <- gseaplot(GSEA.KEGG, geneSetID = i, by = "preranked", title = GSEA.KEGG$Description[i])
  p3 <- gseaplot(GSEA.KEGG, geneSetID = i, title = GSEA.KEGG$Description[1])
  p <- cowplot::plot_grid(p1, p2, p3, ncol=1, labels=LETTERS[1:3])
  pw <- GSEA.KEGG$Description[i]
  cat(sprintf("#### %s \n", pw))
  print(p)
  cat("\n \n")
}
```

```{r echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.width=28, fig.height=20}
```
